<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Several implementations of a novel Bayesian hierarchical statistical model of nucleotide
	recoding RNA-seq experiments (NR-seq; TimeLapse-seq, SLAM-seq, TUC-seq, etc.)
	for analyzing and comparing NR-seq datasets (see Vock and Simon (2022) &lt;doi:10.1101/2022.09.02.505697&gt;).
	NR-seq is a powerful extension of RNA-seq that provides information about the kinetics
	of RNA metabolism (e.g., RNA degradation rate constants), which is notably lacking
	in standard RNA-seq data. The statistical model makes maximal use of these high-throughput
	datasets by sharing information across transcripts to significantly improve
	uncertainty quantification and increase statistical power. bakR includes a maximally
	efficient implementation of this model for conservative initial investigations of datasets. bakR
	also provides more highly powered implementations using the probabilistic programming language
	Stan to sample from the full posterior distribution. bakR performs multiple-test
	adjusted statistical inference with the output of these model implementations to
	help biologists separate signal from background. Methods to automatically visualize key
	results and detect batch effects are also provided.">
<title>Analyze and Compare Nucleotide Recoding RNA Sequencing Datasets • bakR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analyze and Compare Nucleotide Recoding RNA Sequencing Datasets">
<meta property="og:description" content="Several implementations of a novel Bayesian hierarchical statistical model of nucleotide
	recoding RNA-seq experiments (NR-seq; TimeLapse-seq, SLAM-seq, TUC-seq, etc.)
	for analyzing and comparing NR-seq datasets (see Vock and Simon (2022) &lt;doi:10.1101/2022.09.02.505697&gt;).
	NR-seq is a powerful extension of RNA-seq that provides information about the kinetics
	of RNA metabolism (e.g., RNA degradation rate constants), which is notably lacking
	in standard RNA-seq data. The statistical model makes maximal use of these high-throughput
	datasets by sharing information across transcripts to significantly improve
	uncertainty quantification and increase statistical power. bakR includes a maximally
	efficient implementation of this model for conservative initial investigations of datasets. bakR
	also provides more highly powered implementations using the probabilistic programming language
	Stan to sample from the full posterior distribution. bakR performs multiple-test
	adjusted statistical inference with the output of these model implementations to
	help biologists separate signal from background. Methods to automatically visualize key
	results and detect batch effects are also provided.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="articles/Further-Analyses.html">Going beyond the defaults with bakR</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="brief-description-of-bakr">Brief Description of bakR<a class="anchor" aria-label="anchor" href="#brief-description-of-bakr"></a>
</h1></div>
<p>bakR (Bayesian analysis of the kinetics of RNA) is an R package for performing differential kinetics analysis with nucleotide recoding high-throughput RNA sequencing (NR-seq) data. Kinetic parameter estimation and statistical testing is compatible with mutational data from any enrichment free NR-seq method (e.g., TimeLapse-seq, SLAM-seq, TUC-seq, etc.).</p>
</div>
<div class="section level1">
<h1 id="why-use-bakr">Why use bakR?<a class="anchor" aria-label="anchor" href="#why-use-bakr"></a>
</h1>
<p>Differential expression analysis of RNA sequencing (RNA-seq) data can identify changes in cellular RNA levels, but cannot determine the kinetic mechanism underlying such changes. Previously, <a href="https://simonlab.yale.edu/research/transcriptome-dynamics/timelapse-chemistry/" class="external-link">our lab</a> and others addressed this shortcoming by developing nucleotide-recoding RNA-seq methods (NR-seq; e.g., TimeLapse-seq) to quantify changes in RNA synthesis and degradation kinetics. While advanced statistical models implemented in user-friendly software (e.g., DESeq2) have ensured the statistical rigor of differential expression analyses, no such tools that facilitate differential kinetic analysis with NR-seq exist. To address this need, we developed bakR, an R package that analyzes and compares NR-seq datasets. Differential kinetics analysis with bakR relies on a Bayesian hierarchical model of NR-seq data to increase statistical power by sharing information across transcripts. bakR outperforms attempts to use single sample analysis tools (e.g., pulseR and GRAND-SLAM) for differential kinetics analysis. Check out <a href="https://www.biorxiv.org/content/10.1101/2022.09.02.505697v1" class="external-link">our preprint</a> to learn more about the model and its extensive validation!</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>bakR is now available on CRAN! If you are using a Mac OS (and any version of R) or Windows and R version 4.2.1, then that means you don’t need to configure a C++ compiler to install and use bakR. Eventually, this will also hold for users of the Windows OS and any version of R, CRAN just needs time to build the binaries for older R versions and R-devel. Those not on a Mac OS or Windows + R4.2.1 need to first properly configure a C++ compiler; see the next paragraph for details and links describing how to do that. In either case, once you (and your compiler if necessary) are ready, bakR can be installed as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">install.packages</span>(<span class="st">"bakR"</span>) </span></code></pre></div>
<p>To install the newest version of bakR from Github, you need to have a C++ compiler configured to rstan’s (the R interface to the probabilistic programming language <a href="https://mc-stan.org/" class="external-link">Stan</a> that bakR uses on the backend) liking. The best way to do this is to follow the Stan team’s <a href="https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started" class="external-link">helpful documentation</a> on installing rstan for your operating system. Once that is complete, you can install bakR as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">install.packages</span>(<span class="st">"devtools"</span>) <span class="co"># if you haven't installed devtools already</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"simonlabcode/bakR"</span>)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h1>
<p>There are currently two vignettes to help get you up to speed with using bakR:</p>
<ol style="list-style-type: decimal">
<li>An introductory vignette (title: Differential Kinetic Analysis with bakR) that walks you through the basic bakR workflow with simulated data.</li>
<li>A second vignette (title: Going beyond the defaults with bakR) that dives into some specific follow-up analyses one can perform after running bakR on a dataset. Currently, this vignette discusses differential synthesis analysis (i.e., identifying differences in RNA synthesis rates by combining bakR with differential expression analysis) and performing analyses without assuming steady-state. More analyses will be added to this vignette as time goes by.</li>
</ol>
<p>All vignettes are available on the <a href="https://simonlabcode.github.io/bakR/index.html">bakR website</a> under the Articles section. <a href="https://github.com/simonlabcode/bakR" class="external-link">Here</a> is the link to the bakR github as well if you need help getting back to the github from the website.</p>
</div>
<div class="section level1">
<h1 id="obtaining-the-necessary-input">Obtaining the Necessary Input<a class="anchor" aria-label="anchor" href="#obtaining-the-necessary-input"></a>
</h1>
<p>As discussed in the introductory vignette, bakR requires data in the form of a so-called “cB”, or counts binomial data frame. Each row of the cB data frame corresponds to a group of reads with identical mutational data, and the columns denote the sample from which the reads came, the feature the reads aligned to, the number of mutations of interest in the reads (e.g., T-to-C mutations), the number of mutable positions (e.g. Ts), and the number of such reads. It is reasonable to wonder “where am I supposed to get this information?” While there are a couple possibilities, perhaps the easiest and most widely applicable is <a href="https://github.com/simonlabcode/bam2bakR" class="external-link">bam2bakR</a>, a Snakemake implementation of the <a href="https://bitbucket.org/mattsimon9/timelapse_pipeline/src/master/" class="external-link">TimeLapse pipeline</a> developed by the Simon lab. bam2bakR takes as input aligned bam files and produces, among other things, the cB file required by bakR. Extensive documentation describing how to get bam2bakR up and running is available on its GitHub repo. Snakemake greatly facilitates running this pipeline on almost any computational infrastructure and bam2bakR uses the conda/mamba package manager to make setting up the necessary dependencies a breeze.</p>
</div>
<div class="section level1">
<h1 id="bug-catching-and-further-questions">Bug Catching and Further Questions<a class="anchor" aria-label="anchor" href="#bug-catching-and-further-questions"></a>
</h1>
<p>Post descriptions of bugs and a simple reproducible example (if possible) in the Issues section of this repo. In fact, you should go to the Issues section with any question you have about bakR, and there are even helpful labels that you can append to your posts to make the nature of your request clear. If you email me (Isaac Vock) with a question/concern/suggestion, I will direct you to the Issues section. If you have basic use questions, I would suggest going through the vignettes linked above. If these do not answer your question, then post your question to Issues.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=bakR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/simonlabcode/bakR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/simonlabcode/bakR/issues/" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing bakR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Isaac Vock <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7178-6886" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
