---
title: "Differential Kinetic Analysis with DynamicSeq"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Getting-Started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
abstract: >
  In a standard RNA-seq experiment, differential expression 
  analysis is incredibly common. The idea is to compare the expression
  of thousands of genes or transcripts between two conditions, which
  I'll refer to as the reference and experimental samples. Once you 
  have compiled a list of genes whose expressions differ significantly
  between the two conditions, a reasonable follow-up question is "what
  caused the change in each of the genes"? The three broad answers to
  that question are 1) Something altered the RNA synthesis rates; 2) Something
  altered the RNA degradation rates (i.e., its stability); or 3) A little bit
  of both 1) and 2). Unfortunately, standard RNA-seq analysis cannot answer this
  mechanistic question, as RNA-seq only tells you about changes in expression level
  and cannot deduce the kinetic cause of those changes. 
  
  
  If the kinetics is of interest to you, then a recently developed solution to 
  this problem is nucleotide recoding RNA-seq, which exists in several fundamentally 
  similar flavors, TimeLapse-seq, SLAM-seq, and TUC-seq. The strategy involves feeding
  cells with a metabolic label (typically 4-thiouridine, a.k.a. s4U), extracting the RNA as you
  normally would for RNA-seq, but then using a unique chemistry to convert the metabolic
  label (e.g., s4U) into something that a sequencer will read as a different base (e.g.,
  a cytosine analog). Thus, labeled and unlabeled reads can be separated bioinformatically
  upon sequencing by the presence of mutations (e.g., U-to-C mutations). 
  
  
  One problem though; the metabolic label must 1st get into cells, and then it must compete 
  with the normal nucleotide for incorporation into RNA. The result of this is that mutation
  rates are typically low (~ 5% in RNA produced during metabolic labeling, meaning that each U
  in a sequencing read has about a 5% chance of being mutated to a C), making simple and 
  definitive separation of labeled and unlabeled reads impossible. The solution: advanced
  statistical modeling implemented by DynamicSeq, an R package for analyze nucleotide recoding
  sequencing data. The analyses implemented in this package will allow you to perform 
  differential kinetic analysis just as easily as you can perform differential expression analysis,
  allowing you to elucidate the mechanisms of regulated gene expression.
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Necessary Setup

Of course you will want to load the DynamicSeq package for
this vignette, but we'll also be using utils, so make sure 
to load (and if necessary, install) that as well. Finally, you'll
want to set the seed so as to ensure the results you get reproduce
those presented in the vignette. Additionally, you will need to install
[Stan](https://mc-stan.org/), which is the statistical modeling software
that DynamicSeq makes use of under the hood for the heavier, higher performance
models.

```{r setup}
library(DynamicSeq)
library(utils)
set.seed(123)
```

## Step 1: Creating a DynamicSeqData Object

The 1st step to using DynamicSeq is to create a DynamicSeqData
object. A DynamicSeqData object
consists of two components: a cB dataframe and a metadf dataframe.
cB stands for counts binomial and contains all of the information about
mutations seen in sequencing reads in each sample sequenced. metadf stands
for metadata data frame and contains important information about the
experimental details of each sample (i.e., how long the metabolic label
feed was, which samples are reference samples, and which are experimental
samples). 


Lets get an idea for what a cB should look like 1st.

```{r, results = "hide"}
# Load small example cB; will be called cB_small in global environment
data("cB_example")

# Observe contents of cB
head(cB_small)

```

```{r, echo = FALSE, warning = FALSE}

knitr::kable(head(cB_small), "pipe")

```

One key aspect of the cB dataframe to keep in mind is that each row corresponds to a set of reads providing identical data. To help that make more sense,
and to understand what information the cB tracks, let's go through the contents
of each column:

1. **sample**: This is the name of the sample from which the reads described in this
row originated. Usually this will be some sort of character vector that you have
defined.
1. **TC**: Number of U-to-C mutations (assuming using s4U as the metabolic label). Called TC because technically the data is T-to-C mutations in reverse-transcribed RNA
1. **nT**: Number of Ts (Us in the RNA) in the sequencing read(s)
1. **XF**: Name of the feature (gene, exon, etc.) that the read comes from. Kinetic parameters will be estimated for each feature, so this should represent the functional unit who's metabolic kinetics you are interested in. 
    * **XF** stands for "exonic feature", since in most cases, you        should only be considering reads that map to definitively       exonic locations. Introns are typically rapidly turned          over and thus will be highly labeled species that will          bias estimates of RNA stability.
1. **n**: Number of reads with identical data for the other 4 columns. For example, n = 2 in the 4th row means that there are 2 reads from sample JS181101 that mapped to **XF** EPB41L1, had 0 T to C mutations, and had 14 Ts.

The structure of the cB file most closely reflects the output of a pipeline for analyze nucleotide recoding sequencing data developed by the Simon lab (available [here](https://bitbucket.org/mattsimon9/timelapse_pipeline/src/master/)), but consists of data provided by all published pipelines (e.g., SlamDunk). 

Now let's look at the corresponding metadf dataframe:

```{r, results = "hide"}
# Load metadf data frame; will be loaded as metadf in global environment
data("metadf_example")

# Print the transpose of metadf
# Rows will be the columns and columns will be the rows
print(t(metadf))

```

```{r, echo = FALSE, warning = FALSE}
knitr::kable(t(metadf))

```

The transpose of metadf is displayed above to save space (so the columns depicted are the rows of metadf, and vice versa). Metadf has 
significantly less information than the cB data frame, but it's crucial information. One extremely important feature of metadf is its row names (column names as depicted above). Examine the row names and compare them to the sample names

```{r}
# metadf row names
print(rownames(metadf))


# cB sample names
print(unique(cB_small$sample))

```

They're the same, and they HAVE TO BE!! Each row of metadf corresponds to a sample, and the order in which the sample names appear as the row names of metadf **MUST** correspond to the order they appear in cB sample column. 

Keeping in mind that each row represents a sample, the content of the columns are:

1. **tl**: The length of the metabolic labeling feed. This can be in any units (the simulated data puts it in terms of minutes), but **if no s4U fed to a sample, tl must be 0 for that sample**. -s4U controls play an important role in all of the statistical models, so this is crucial to remember.
1. **Exp_ID**: Stands for "Experimental ID" and is a numerical ID that denotes which samples are from the reference condition and which experimental samples belong to the same "treatment". In this simulated dataset, there are two conditions, a reference condition (samples 1, 2, 3, and 7) and a single experimental condition (samples 4, 5, 6, and 8). **The reference condition must be given numerical ID 1** and **the experimental conditions must be given sequential numerical IDs starting from 2** (i.e., 2, 3, 4, ...)

Once you have these two data frames correctly constructed, you can create a DynamicSeqData object in one line of code.

```{r}
# Create DynamicSeqData object
DynData <- DynamicSeqData(cB_small, metadf)

```
