<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="cBprocess creates the data structures necessary to analyze nucleotide recoding RNA-seq data with any of the
statistical model implementations in bakRFit. The input to cBprocess must be an object of class
bakRData. The output can contain data passable to fast_analysis (if Fast == TRUE), TL_stan with StanFit = TRUE if
Stan == TRUE, or both."><title>Curate data in bakRData object for statistical modeling — cBprocess • bakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Curate data in bakRData object for statistical modeling — cBprocess"><meta property="og:description" content="cBprocess creates the data structures necessary to analyze nucleotide recoding RNA-seq data with any of the
statistical model implementations in bakRFit. The input to cBprocess must be an object of class
bakRData. The output can contain data passable to fast_analysis (if Fast == TRUE), TL_stan with StanFit = TRUE if
Stan == TRUE, or both."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/Further-Analyses.html">Going beyond the defaults with bakR</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Curate data in bakRData object for statistical modeling</h1>
      
      <div class="d-none name"><code>cBprocess.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>cBprocess</code> creates the data structures necessary to analyze nucleotide recoding RNA-seq data with any of the
statistical model implementations in <code>bakRFit</code>. The input to <code>cBprocess</code> must be an object of class
<code>bakRData</code>. The output can contain data passable to <code>fast_analysis</code> (if Fast == TRUE), <code>TL_stan</code> with StanFit = TRUE if
Stan == TRUE, or both.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cBprocess</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  high_p <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  totcut <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  Ucut <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  AvgU <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  Stan <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Fast <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  FOI <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  concat <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>obj</dt>
<dd><p>An object of class bakRData</p></dd>


<dt>high_p</dt>
<dd><p>Numeric; Any transcripts with a mutation rate (number of mutations / number of Ts in reads) higher than this in any no s4U control
samples are filtered out</p></dd>


<dt>totcut</dt>
<dd><p>Numeric; Any transcripts with less than this number of sequencing reads in any sample are filtered out</p></dd>


<dt>Ucut</dt>
<dd><p>Numeric; All transcripts must have a fraction of reads with 2 or less Us less than this cutoff in all samples</p></dd>


<dt>AvgU</dt>
<dd><p>Numeric; All transcripts must have an average number of Us greater than this cutoff in all samples</p></dd>


<dt>Stan</dt>
<dd><p>Boolean; if TRUE, then data_list that can be passed to Stan is curated</p></dd>


<dt>Fast</dt>
<dd><p>Boolean; if TRUE, then dataframe that can be passed to fast_analysis() is curated</p></dd>


<dt>FOI</dt>
<dd><p>Features of interest; character vector containing names of features to analyze</p></dd>


<dt>concat</dt>
<dd><p>Boolean; If TRUE, FOI is concatenated with output of reliableFeatures</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>returns list of objects that can be passed to <code>TL_stan</code> and/or <code>fast_analysis</code>. Those objects are:</p><ul><li><p>Stan_data; list that can be passed to <code>TL_stan</code> with Hybrid_Fit = FALSE. Consistents of metadata as well as data that
Stan will analyze. Data to be analyzed consists of equal length vectors. The contents of Stan_data are:</p><ul><li><p>NE; Number of datapoints for Stan to analyze (NE = Number of Elements)</p></li>
<li><p>NF; Number of features in dataset</p></li>
<li><p>TP; Numerical indicator of s4U feed (0 = no s4U feed, 1 = s4U fed)</p></li>
<li><p>FE; Numerical indicator of feature</p></li>
<li><p>num_mut; Number of U-to-C mutations observed in a particular set of reads</p></li>
<li><p>MT; Numerical indicator of experimental condition (Exp_ID from metadf)</p></li>
<li><p>nMT; Number of experimental conditions</p></li>
<li><p>R; Numerical indicator of replicate</p></li>
<li><p>nrep; Number of replicates (analysis requires same number of replicates of all conditions)</p></li>
<li><p>num_obs; Number of reads with identical data (number of mutations, feature of origin, and sample of origin)</p></li>
<li><p>tl; Vector of label times for each experimental condition</p></li>
<li><p>U_cont; Log2-fold-difference in U-content for a feature in a sample relative to average U-content for that sample</p></li>
<li><p>Avg_Reads; Standardized log10(average read counts) for a particular feature in a particular condition, averaged over
replicates</p></li>
<li><p>Avg_Reads_natural; Unstandardized average read counts for a particular feature in a particular condition, averaged over
replicates. Used for <code>plotMA</code></p></li>
<li><p>sdf; Dataframe that maps numerical feature ID to original feature name. Also has read depth information</p></li>
<li><p>sample_lookup; Lookup table relating MT and R to the original sample name</p></li>
</ul></li>
<li><p>Fast_df; A data frame that can be passed to <code>fast_analysis</code>. The contents of Fast_df are:</p><ul><li><p>sample; Original sample name</p></li>
<li><p>XF; Original feature name</p></li>
<li><p>TC; Number of T to C mutations</p></li>
<li><p>nT; Number of Ts in read</p></li>
<li><p>n; Number of identical observations</p></li>
<li><p>fnum; Numerical indicator of feature</p></li>
<li><p>type; Numerical indicator of s4U feed (0 = no s4U feed, 1 = s4U fed)</p></li>
<li><p>mut; Numerical indicator of experimental condition (Exp_ID from metadf)</p></li>
<li><p>reps; Numerical indicator of replicate</p></li>
</ul></li>
<li><p>Count_Matrix; A matrix with read count information. Each column represents a sample and each row represents a feature.
Each entry is the raw number of read counts mapping to a particular feature in a particular sample. Column names are the corresponding
sample names and row names are the corresponding feature names.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The 1st step executed by <code>cBprocess</code> is to find the names of features which are deemed "reliable". A reliable feature is one with
sufficient read coverage in every single sample (i.e., &gt; totcut reads in all samples) and limited mutation content in all -s4U
control samples (i.e., &lt; high_p mutation rate in all samples lacking s4U feeds). This is done with a call to <code>reliableFeatures</code>.
The 2nd step is to extract only reliableFeatures from the cB dataframe in the <code>bakRData</code> object. During this process, a numerical
ID is given to each reliableFeature, with the numerical ID corresponding to the order in which each feature is found in the original cB
(this might typically be alphabetical order).</p>
<p>The 3rd step is to prepare a dataframe where each row corresponds to a set of n identical reads (that is they come from the same sample
and have the same number of mutations and Us). Part of this process involves assigning an arbitrary numerical ID to each replicate in each
experimental condition. The numerical ID will correspond to the order the sample appears in metadf. The outcome of this step is multiple
dataframes with variable information content. These include a dataframe with information about read counts in each sample, one which logs
the U-contents of each feature, one which is compatible with <code>fast_analysis</code> and thus groups reads by their number of mutations as
well as their number of Us, and one which is compatible with <code>TL_stan</code> with StanFit == TRUE and thus groups ready by only their number
of mutations. At the end of this step, two other smaller data structures are created, one which is an average count matrix (a count matrix
where the ith row and jth column corresponds to the average number of reads mappin to feature i in experimental condition j, averaged over
all replicates) and the other which is a sample lookup table that relates the numerical experimental and replicate IDs to the original
sample name.</p>
<p>If FOI is non-null and concat == TRUE, the features listed in FOI will be included in the list of reliable features that make it past
filtering. If FOI is non-null and concat == FALSE, the features listed in FOI will be the only reliable features that make it past filtering.
If FOI is null and concat == FALSE, an error will be thrown.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

