<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Correcting for metabolic labeling induced RNA dropout — CorrectDropout • bakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Correcting for metabolic labeling induced RNA dropout — CorrectDropout"><meta name="description" content="Dropout is the name given to a phenomenon originally identified by our lab and
further detailed in two independent publications (Zimmer et al. (2023),
and  Berg et al. (2023)).
Dropout is the under-representation of reads from RNA containing metabolic label
(4-thiouridine or 6-thioguanidine most commonly). Loss of 4-thiouridine (s4U)
containing RNA on plastic surfaces and RT dropoff caused by
modifications on s4U introduced by recoding chemistry have been attributed as the likely
causes of this phenomenon. While protocols can be altered in ways to drastically reduce this
source of dropout, you may still have datasets that you want to analyze with bakR collected
with suboptimal handling. That is where CorrectDropout comes in."><meta property="og:description" content="Dropout is the name given to a phenomenon originally identified by our lab and
further detailed in two independent publications (Zimmer et al. (2023),
and  Berg et al. (2023)).
Dropout is the under-representation of reads from RNA containing metabolic label
(4-thiouridine or 6-thioguanidine most commonly). Loss of 4-thiouridine (s4U)
containing RNA on plastic surfaces and RT dropoff caused by
modifications on s4U introduced by recoding chemistry have been attributed as the likely
causes of this phenomenon. While protocols can be altered in ways to drastically reduce this
source of dropout, you may still have datasets that you want to analyze with bakR collected
with suboptimal handling. That is where CorrectDropout comes in."><meta property="og:image" content="https://simonlabcode.github.io/bakR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a></li>
    <li><a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a></li>
    <li><a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a></li>
    <li><a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a></li>
    <li><a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a></li>
    <li><a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Correcting for metabolic labeling induced RNA dropout</h1>
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/master/R/Dropout_correction.R" class="external-link"><code>R/Dropout_correction.R</code></a></small>
      <div class="d-none name"><code>CorrectDropout.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Dropout is the name given to a phenomenon originally identified by our lab and
further detailed in two independent publications (<a href="https://www.biorxiv.org/content/10.1101/2023.05.24.542133v1" class="external-link">Zimmer et al. (2023)</a>,
and <a href="https://www.biorxiv.org/content/10.1101/2023.04.21.537786v1" class="external-link"> Berg et al. (2023)</a>).
Dropout is the under-representation of reads from RNA containing metabolic label
(4-thiouridine or 6-thioguanidine most commonly). Loss of 4-thiouridine (s4U)
containing RNA on plastic surfaces and RT dropoff caused by
modifications on s4U introduced by recoding chemistry have been attributed as the likely
causes of this phenomenon. While protocols can be altered in ways to drastically reduce this
source of dropout, you may still have datasets that you want to analyze with bakR collected
with suboptimal handling. That is where <code>CorrectDropout</code> comes in.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CorrectDropout</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  scale_init <span class="op">=</span> <span class="fl">1.05</span>,</span>
<span>  pdo_init <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  recalc_uncertainty <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>bakRFit object</p></dd>


<dt id="arg-scale-init">scale_init<a class="anchor" aria-label="anchor" href="#arg-scale-init"></a></dt>
<dd><p>Numeric; initial estimate for -s4U/+s4U scale factor. This is the factor
difference in RPM normalized read counts for completely unlabeled transcripts (i.e., highly stable
transcript) between the +s4U and -s4U samples.</p></dd>


<dt id="arg-pdo-init">pdo_init<a class="anchor" aria-label="anchor" href="#arg-pdo-init"></a></dt>
<dd><p>Numeric; initial estimtae for the dropout rate. This is the probability
that an s4U labeled RNA molecule is lost during library prepartion.</p></dd>


<dt id="arg-recalc-uncertainty">recalc_uncertainty<a class="anchor" aria-label="anchor" href="#arg-recalc-uncertainty"></a></dt>
<dd><p>Logical; if TRUE, then fraction new uncertainty is recalculated
using adjusted fn and a simple binomial model of estimate uncertainty. This will provide a
slight underestimate of the fn uncertainty, but will be far less biased for low coverage features,
or for samples with low pnews.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional (optional) parameters to be passed to <code><a href="https://rdrr.io/r/stats/nls.html" class="external-link">stats::nls()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>bakRFit</code> or <code>bakRFnFit</code> object (same type as was passed in). Fraction new estimates and read counts
in <code>Fast_Fit$Fn_Estimates</code> and (in the case of a <code>bakRFnFit</code> input) <code>Data_lists$Fn_Est</code>are dropout corrected.
A count matrix with corrected read counts (<code>Data_lists$Count_Matrix_corrected</code>) is also output, along with a
data frame with information about the dropout rate estimated for each sample (<code>Data_lists$Dropout_df</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>CorrectDropout</code> estimates the percentage of 4-thiouridine containing RNA
that was lost during library preparation (pdo). It then uses this estimate of pdo
to correct fraction new estimates and read counts. Both corrections are analytically
derived from a rigorous generative model of NR-seq data. Importantly, the read count
correction preserves the total library size to avoid artificially inflating read counts.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Simulate data for 500 genes and 2 replicates with 40% dropout</span></span></span>
<span class="r-in"><span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="Simulate_relative_bakRData.html">Simulate_relative_bakRData</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">100000</span>, nreps <span class="op">=</span> <span class="fl">2</span>, p_do <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Fit data with fast implementation</span></span></span>
<span class="r-in"><span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="va">bakRData</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding reliable Features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering out unwanted or unreliable features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating pnew with likelihood maximization</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating unlabeled mutation rate with -s4U data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimated pnews and polds for each sample are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 4 × 4</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Groups:   mut [2]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     mut  reps   pnew    pold</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     1     1 0.050<span style="text-decoration: underline;">1</span> 0.001<span style="text-decoration: underline;">02</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">5</span> 0.001<span style="text-decoration: underline;">02</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span>     2     1 0.049<span style="text-decoration: underline;">8</span> 0.001<span style="text-decoration: underline;">02</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span>     2     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">02</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating fraction labeled</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating per replicate uncertainties</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating read count-variance relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Averaging replicate data and regularizing estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assessing statistical significance</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> All done! Run QC_checks() on your bakRFit object to assess the </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>             quality of your data and get recommendations for next steps.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Correct for dropout</span></span></span>
<span class="r-in"><span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu">CorrectDropout</span><span class="op">(</span><span class="va">Fit</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimated rates of dropout are:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Exp_ID Replicate       pdo</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1      1         1 0.4169686</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2      1         2 0.2494506</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3      2         1 0.0000000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4      2         2 0.0000000</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mapping sample name to sample characteristics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering out low coverage features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating read count-variance relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Averaging replicate data and regularizing estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assessing statistical significance</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> All done! Run QC_checks() on your bakRFit object to assess the </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>             quality of your data and get recommendations for next steps.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

