<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="sim_bakRData simulates a bakRData object. It's output also includes the simulated
values of all kinetic parameters of interest. Only the number of genes (ngene) has to be set by the
user, but an extensive list of additional parameters can be adjusted."><title>Simulating nucleotide recoding data — sim_bakRData • bakR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating nucleotide recoding data — sim_bakRData"><meta property="og:description" content="sim_bakRData simulates a bakRData object. It's output also includes the simulated
values of all kinetic parameters of interest. Only the number of genes (ngene) has to be set by the
user, but an extensive list of additional parameters can be adjusted."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/Further-Analyses.html">Going beyond the defaults with bakR</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulating nucleotide recoding data</h1>
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/R/Simulations.R" class="external-link"><code>R/Simulations.R</code></a></small>
      <div class="d-none name"><code>sim_bakRData.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>sim_bakRData</code> simulates a <code>bakRData</code> object. It's output also includes the simulated
values of all kinetic parameters of interest. Only the number of genes (<code>ngene</code>) has to be set by the
user, but an extensive list of additional parameters can be adjusted.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sim_bakRData</span><span class="op">(</span></span>
<span>  <span class="va">ngene</span>,</span>
<span>  num_conds <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>  nreps <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  eff_sd <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  eff_mean <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  fn_mean <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  fn_sd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  kslog_c <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  kslog_sd <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  tl <span class="op">=</span> <span class="fl">60</span>,</span>
<span>  p_new <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  p_old <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  read_lengths <span class="op">=</span> <span class="fl">200L</span>,</span>
<span>  p_do <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  noise_deg_a <span class="op">=</span> <span class="op">-</span><span class="fl">0.3</span>,</span>
<span>  noise_deg_b <span class="op">=</span> <span class="op">-</span><span class="fl">1.5</span>,</span>
<span>  noise_synth <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  sd_rep <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  low_L2FC_ks <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  high_L2FC_ks <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  num_kd_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0L</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">ngene</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>, times <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">num_conds</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  num_ks_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0L</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">num_conds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  scale_factor <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  sim_read_counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  a1 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  a0 <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  nreads <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  beta <span class="op">=</span> <span class="fl">75</span>,</span>
<span>  STL <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  STL_len <span class="op">=</span> <span class="fl">40</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ngene</dt>
<dd><p>Number of genes to simulate data for</p></dd>


<dt>num_conds</dt>
<dd><p>Number of experimental conditions (including the reference condition) to simulate</p></dd>


<dt>nreps</dt>
<dd><p>Number of replicates to simulate</p></dd>


<dt>eff_sd</dt>
<dd><p>Effect size; more specifically, the standard deviation of the normal distribution from which non-zero
changes in logit(fraction new) are pulled from.</p></dd>


<dt>eff_mean</dt>
<dd><p>Effect size mean; mean of normal distribution from which non-zero changes in logit(fraction new) are pulled from.
Note, setting this to 0 does not mean that some of the significant effect sizes will be 0, as any exact integer is impossible
to draw from a continuous random number generator. Setting this to 0 just means that there is symmetric stabilization and destabilization</p></dd>


<dt>fn_mean</dt>
<dd><p>Mean of fraction news of simulated transcripts in reference condition. The fraction of RNA from each transcript that is
s4U labeled (new) is drawn from a normal distribution with this mean</p></dd>


<dt>fn_sd</dt>
<dd><p>Standard deviation of fraction news of simulated transcripts in reference condition. The fraction of RNA from each transcript that
is s4U labeled (new) is drawn from a normal distribution with this sd</p></dd>


<dt>kslog_c</dt>
<dd><p>Synthesis rate constants will be drawn from a lognormal distribution with meanlog = <code>kslog_c</code> - mean(log(kd_mean)) where kd_mean
is determined from the fraction new simulated for each gene as well as the label time (<code>tl</code>).</p></dd>


<dt>kslog_sd</dt>
<dd><p>Synthesis rate lognormal standard deviation; see kslog_c documentation for details</p></dd>


<dt>tl</dt>
<dd><p>s4U label feed time</p></dd>


<dt>p_new</dt>
<dd><p>s4U induced mutation rate. Can be a vector of length num_conds</p></dd>


<dt>p_old</dt>
<dd><p>background mutation rate</p></dd>


<dt>read_lengths</dt>
<dd><p>Total read length for each sequencing read (e.g., 200 means PE100 reads)</p></dd>


<dt>p_do</dt>
<dd><p>Rate at which s4U containing reads are lost due to dropout; must be between 0 and 1</p></dd>


<dt>noise_deg_a</dt>
<dd><p>Slope of trend relating log10(standardized read counts) to log(replicate variability)</p></dd>


<dt>noise_deg_b</dt>
<dd><p>Intercept of trend relating log10(standardized read counts) to log(replicate variability)</p></dd>


<dt>noise_synth</dt>
<dd><p>Homoskedastic variability of L2FC(ksyn)</p></dd>


<dt>sd_rep</dt>
<dd><p>Variance of lognormal distribution from which replicate variability is drawn</p></dd>


<dt>low_L2FC_ks</dt>
<dd><p>Most negative L2FC(ksyn) that can be simulated</p></dd>


<dt>high_L2FC_ks</dt>
<dd><p>Most positive L2FC(ksyn) that can be simulated</p></dd>


<dt>num_kd_DE</dt>
<dd><p>Vector where each element represents the number of genes that show a significant change in stability relative
to the reference. 1st entry must be 0 by definition (since relative to the reference the reference sample is unchanged)</p></dd>


<dt>num_ks_DE</dt>
<dd><p>Same as num_kd_DE but for significant changes in synthesis rates.</p></dd>


<dt>scale_factor</dt>
<dd><p>Factor relating RNA concentration (in arbitrary units) to average number of read counts</p></dd>


<dt>sim_read_counts</dt>
<dd><p>Logical; if TRUE, read counts are simulated as coming from a heterodisperse negative binomial distribution</p></dd>


<dt>a1</dt>
<dd><p>Heterodispersion 1/reads dependence parameter</p></dd>


<dt>a0</dt>
<dd><p>High read depth limit of negative binomial dispersion parameter</p></dd>


<dt>nreads</dt>
<dd><p>Number of reads simulated if sim_read_counts is FALSE</p></dd>


<dt>alpha</dt>
<dd><p>shape1 parameter of the beta distribution from which U-contents (probability that a nucleotide in a read from a transcript is a U) are
drawn for each gene.</p></dd>


<dt>beta</dt>
<dd><p>shape2 parameter of the beta distribution from which U-contents (probability that a nucleotide in a read from a transcript is a U) are
drawn for each gene.</p></dd>


<dt>STL</dt>
<dd><p>logical; if TRUE, simulation is of STL-seq rather than a standard TL-seq experiment. The two big changes are that a short read lenght is required
(&lt; 60 nt) and that every read for a particular feature will have the same number of Us. Only one read length is simulated for simplicity.</p></dd>


<dt>STL_len</dt>
<dd><p>Average length of simulated STL-seq length. Since Pol II typically pauses about 20-60 bases
from the promoter, this should be around 40</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list containing a simulated <code>bakRData</code> object as well as a list of simulated kinetic parameters of interest.
The contents of the latter list are:</p><ul><li><p>Effect_sim; Dataframe meant to mimic formatting of Effect_df that are part of <code>TL_stan</code> and <code>fast_analysis</code> output.</p></li>
<li><p>Fn_mean_sim; Dataframe meant to mimic formatting of Regularized_ests that is part of <code>fast_analysis</code> output. Contains information
about the true fraction new simulated in each condition (the mean of the normal distribution from which replicate fraction news are simulated)</p></li>
<li><p>Fn_rep_sim; Dataframe meant to mimic formatting of Fn_Estimates that is part of <code>fast_analysis</code> output. Contains information
about the fraction new simulated for each feature in each replicate of each condition.</p></li>
<li><p>L2FC_ks_mean; The true L2FC(ksyn) for each feature in each experimental condition. The ith column corresponds to the L2FC(ksyn) when comparing
the ith condition to the reference condition (defined as the 1st condition) so the 1st column is always all 0s</p></li>
<li><p>RNA_conc; The average number of normalized read counts expected for each feature in each sample.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>sim_bakRData</code> simulates a <code>bakRData</code> object using an extensive list of
adjustable parameters. Average RNA kinetic parameters are drawn from biologically inspired
distributions. Replicate variability is simulated by drawing each replicate and feature's
fraction labeled (aka fraction new) from a Logit-Normal distribution with a heteroskedastic
variance term with average magnitude given by a read count vs. variance relationship, and each
replicate and feature's ksyn from a homoskedastic lognormal distribution. Read counts
can either be set to the same value for all simulated features or can be simulated according to
a heterodisperse negative binomial distribution.</p>
<p>The number of Us in each sequencing read is drawn from a binomial distribution with number of trials
equal to the read length and probability of each nucleotide being a U of 0.25. Each read is assigned to the
labeled or unlabeled population according to a Bernoulli distribution with p = fraction new. The number of
mutations in each read are then drawn from one of two binomial distributions; if the read is assigned to the
labeled population, the number of mutations are drawn from a binomial distribution with number of trials equal
to the number of Us and probability of mutation = <code>p_new</code>; if the read is assigned to the unlabeled population,
the number of mutations is instead drawn from a binomial distribution with the same number of trials but with the probability
of mutation = <code>p_old</code>.</p>
<p>Simulated read counts should be treated as if they are spike-in and RPKM normalized, so the same scale factor can be applied
to each sample when comparing the sequencing reads (like if you are performing differential expression analysis).</p>
<p>Function to simulate a bakRData object according to a heteroskedastic beta-binomial model
of the fraction new.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

