<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bakR">
<title>GRAND-SLAM output/fn estimates as bakR input • bakR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GRAND-SLAM output/fn estimates as bakR input">
<meta property="og:description" content="bakR">
<meta property="og:image" content="https://simonlabcode.github.io/bakR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a>
    <a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a>
    <a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a>
    <a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a>
    <a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a>
    <a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GRAND-SLAM output/fn estimates as bakR input</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/vignettes/bakR-Fn.Rmd" class="external-link"><code>vignettes/bakR-Fn.Rmd</code></a></small>
      <div class="d-none name"><code>bakR-Fn.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>As of bakR version 1.0.0, you can now provide a table of
      fraction new estimates as input. Tools like GRAND-SLAM provide
      such tables, though you can also use your own pipeline for
      estimating the fraction of reads that are new in your NR-seq
      experiments. In this vigngette, I will provide a brief
      introduction to how to use this type of alternative input in bakR.
      An Appendix at the end of this vignette will go deeper into some
      specifics regarding input data structure and analysis
      strategies.</p>
    </div>
    
<div class="section level2">
<h2 id="necessary-setup">Necessary Setup<a class="anchor" aria-label="anchor" href="#necessary-setup"></a>
</h2>
<p>Make sure you have version 1.0.0 of <a href="https://simonlabcode.github.io/bakR/index.html">bakR</a>
installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-1-creating-a-bakrfndata-object">Step 1: Creating a bakRFnData Object<a class="anchor" aria-label="anchor" href="#step-1-creating-a-bakrfndata-object"></a>
</h2>
<p>The 1st step to using bakR with fraction new estimates as input is to
create a bakRFnData object. A bakRFnData object consists of two
components: a fns data frame and a metadf data frame. fns stands for
“fraction new estimates” and contains information regarding the
estimates for the fraction of sequencing reads from a given feature that
were new in a given sample. metadf stands for metadata data frame and
contains important information about the experimental details of each
sample (i.e., how long the metabolic label feed was, which samples are
reference samples, and which are experimental samples). Examples of what
these data structures are available via calls to
<code>data("fns")</code> and <code>data("metadf")</code>, for the fns
data frame and metadf respectively.</p>
<p>A fns data frame consists of rows corresponding to fraction new
estimate information for a given feature. In particular, there are 4
required (and one optional) columns in the fns data frame:</p>
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong>: This is the name of the sample from which
the reads described in this row originated. Usually this will be some
sort of character vector that you have defined.</li>
<li>
<strong>XF</strong>: Name of the feature (gene, exon, etc.) that the
read comes from. Kinetic parameters will be estimated for each feature,
so this should represent the functional unit who’s metabolic kinetics
you are interested in.
<ul>
<li>
<strong>XF</strong> stands for “exonic feature”, since in most cases
you should only be considering reads that map to definitively exonic
locations. Introns are typically rapidly turned over and thus will be
highly labeled species that will bias estimates of mature transcript
stability.</li>
</ul>
</li>
<li>
<strong>fn</strong>: Estimate of the fraction new</li>
<li>
<strong>se</strong> (optional): Uncertainty in fraction new
estimate</li>
<li>
<strong>n</strong>: Number of reads from the feature/sample
combination represented in this row.</li>
</ol>
<p>The metadf data frame is described in both the “bakR for people in a
hurry” and “Differential kinetic analysis with bakR” vignettes.</p>
<p>One route by which to obtain the fns data frame is via <a href="https://academic.oup.com/bioinformatics/article/34/13/i218/5045735" class="external-link">GRAND-SLAM</a>,
an efficient and user friendly tool developed by the Erhard lab which
implements a binomial mixture model <a href="https://www.nature.com/articles/nmeth.4582#Sec9" class="external-link">originally
described</a> by our lab. GRAND-SLAM outputs a table named
<em>run_name</em>.tsv (where <em>run_name</em> is whatever you specified
when running GRAND-SLAM), which can be quickly converted to a fns data
frame. You can then create a bakRFnData object as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load GRAND-SLAM table and metadf</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GS_table"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"metadf"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create fns data frame from GRAND-SLAM output</span></span>
<span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GSprocessing.html">GSprocessing</a></span><span class="op">(</span><span class="va">GS_table</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Create bakRFnData object</span></span>
<span><span class="va">bfndo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFnData.html">bakRFnData</a></span><span class="op">(</span><span class="va">fns</span>, <span class="va">metadf</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-fitting-the-efficient-model">Step 2: Fitting the Efficient Model<a class="anchor" aria-label="anchor" href="#step-2-fitting-the-efficient-model"></a>
</h2>
<p>After creating the bakRFnData object, you must first run bakR’s most
efficient implementation (the MLE implementation from the bakR
manuscript).:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run the efficient model</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bfndo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mapping sample name to sample characteristics</span></span>
<span><span class="co">#&gt; Filtering out low coverage features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>bakRFit() is used here as a wrapper for two functions in bakR:
fn_process() and fast_analysis(). For more details on what these
functions do, run <code><a href="../reference/fn_process.html">?fn_process</a></code> or
<code>fast_analysis</code>.</p>
<p>When using fraction new estimates as input, only one of the two more
highly powered implementations may be used (the Hybrid implementation
from the bakR manuscript). You can run this as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run Hybrid model (This might take several minutes to run)</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">Fit</span>, HybridFit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The Fit objects contain lists pertaining to the fits of each of the
models. The possible contents include:</p>
<ol style="list-style-type: decimal">
<li>Fast_Fit: Result of the initial fitting of the bakRData object. To
learn more about its contents, see ?fast_analysis()</li>
<li>Data_lists: Processed data that can be passed to the statistical
models</li>
<li>Hybrid_Fit: Result of running bakRFit() on the Fit object with
HybridFit = TRUE. To learn more about its contents, see ?TL_stan().</li>
</ol>
</div>
<div class="section level2">
<h2 id="step-3-visualizing-the-results">Step 3: Visualizing the Results<a class="anchor" aria-label="anchor" href="#step-3-visualizing-the-results"></a>
</h2>
<p>bakR provides a variety of easy to use functions for beginning to
investigate your data. The visualizations are particularly aimed at
revealing trends in RNA stabilization or destabilization. These include
MA plots:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## MA Plot with Fast Fit</span></span>
<span><span class="fu">bakR</span><span class="fu">::</span><span class="fu"><a href="../reference/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">Fit</span>, Model <span class="op">=</span> <span class="st">"MLE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Fn_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Volcano plots:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Volcano Plot with Fast Fit; significance assessed relative to an FDR control of 0.05</span></span>
<span><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Fn_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>and PCA plots:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2D PCA plot with replicate fraction news</span></span>
<span><span class="fu"><a href="../reference/FnPCA2.html">FnPCA2</a></span><span class="op">(</span><span class="va">Fit</span>, Model <span class="op">=</span> <span class="st">"MLE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Fn_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="step-4-where-to-go-from-here">Step 4: Where to go from here<a class="anchor" aria-label="anchor" href="#step-4-where-to-go-from-here"></a>
</h2>
<p>This vignette provides the minimal amount of information to get up
and running with bakR using fraction new estimates as input. If you
would like a more thorough discussion of each step of this process,
check out the long form version of the intro to bakR vignette
(“Differential kinetic analysis with bakR”). In addition, there are a
number of other vignettes that cover various topics not discussed in
these intro vignettes:</p>
<ol style="list-style-type: decimal">
<li>Differential synthesis rate analysis</li>
<li>Correcting for loss of s<sup>4</sup>U RNA during library preparation
(a.k.a. dropout correction)</li>
<li>Troubleshooting analyses of NR-seq datasets</li>
<li>Dissecting the mechanisms of gene expression regulation while
relaxing the steady-state assumption.</li>
</ol>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="a1-input-data">A1: Input data<a class="anchor" aria-label="anchor" href="#a1-input-data"></a>
</h3>
</div>
<div class="section level3">
<h3 id="a2-analysis-options">A2: Analysis options<a class="anchor" aria-label="anchor" href="#a2-analysis-options"></a>
</h3>
</div>
<div class="section level3">
<h3 id="a3-analyzing-non-nr-seq-data">A3: Analyzing non-NR-seq data<a class="anchor" aria-label="anchor" href="#a3-analyzing-non-nr-seq-data"></a>
</h3>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
