<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bakR">
<title>bakR for people in a hurry • bakR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bakR for people in a hurry">
<meta property="og:description" content="bakR">
<meta property="og:image" content="https://simonlabcode.github.io/bakR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a>
    <a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a>
    <a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a>
    <a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a>
    <a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a>
    <a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>bakR for people in a hurry</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/vignettes/bakR-Quickstart.Rmd" class="external-link"><code>vignettes/bakR-Quickstart.Rmd</code></a></small>
      <div class="d-none name"><code>bakR-Quickstart.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>bakR is a tool for comparing nucleotide recoding RNA-seq
      datasets (NR-seq). NR-seq refers to a class of methods (e.g.,
      TimeLapse-seq, SLAM-seq, TUC-seq, etc.) which combine RNA-seq,
      metabolic labeling, and unique metabolic label recoding
      chemistries to assess the kinetics of gene expression. This
      vignette will provide a quick introduction to using bakR. A longer
      version of this vignette (“Differential kinetic analysis with
      bakR”) exists to provide a more in-depth discussion of each step
      of this process.</p>
    </div>
    
<div class="section level2">
<h2 id="necessary-setup">Necessary Setup<a class="anchor" aria-label="anchor" href="#necessary-setup"></a>
</h2>
<p>Install and load bakR to run the code in this vignette; instructions
on how to do so can be found at this <a href="https://simonlabcode.github.io/bakR/index.html">link</a>. Also,
you’ll want to set the seed so as to ensure the results you get
reproduce those presented in the vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-1-creating-a-bakrdata-object">Step 1: Creating a bakRData Object<a class="anchor" aria-label="anchor" href="#step-1-creating-a-bakrdata-object"></a>
</h2>
<p>The 1st step to using bakR is to create a bakRData object. A bakRData
object consists of two components: a cB data frame and a metadf data
frame. cB stands for counts binomial and contains all of the information
about mutations seen in sequencing reads in each sample sequenced.
metadf stands for metadata data frame and contains important information
about the experimental details of each sample (i.e., how long the
metabolic label feed was, which samples are reference samples, and which
are experimental samples). Examples of what these data structures are
available via calls to <code>data("cB_small")</code> and
<code>data("metadf")</code>, for the cB and metadf respectively.</p>
<p>A cB data frame consists of rows corresponding to groups of reads
with identical data, where data corresponds to the values of the
following four variables:</p>
<ol style="list-style-type: decimal">
<li>
<strong>sample</strong>: This is the name of the sample from which
the reads described in this row originated. Usually this will be some
sort of character vector that you have defined.</li>
<li>
<strong>TC</strong>: Number of U-to-C mutations (assuming using
s<sup>4</sup>U as the metabolic label). Called TC because technically
the data is T-to-C mutations in reverse-transcribed RNA.</li>
<li>
<strong>nT</strong>: Number of Ts (Us in the RNA) in the sequencing
read(s)</li>
<li>
<strong>XF</strong>: Name of the feature (gene, exon, etc.) that the
read comes from. Kinetic parameters will be estimated for each feature,
so this should represent the functional unit who’s metabolic kinetics
you are interested in.
<ul>
<li>
<strong>XF</strong> stands for “exonic feature”, since in most cases
you should only be considering reads that map to definitively exonic
locations. Introns are typically rapidly turned over and thus will be
highly labeled species that will bias estimates of mature transcript
stability.</li>
</ul>
</li>
</ol>
<p>The last column of a cB data frame is <strong>n</strong>: Number of
reads with identical data for the other 4 columns.</p>
<p>cB data frames are most easily obtained from the Snakemake
implementation of a pipeline developed by our lab, called bam2bakR
(available <a href="https://github.com/simonlabcode/bam2bakR" class="external-link">here</a>).</p>
<p>The metadf has two columns:</p>
<ol style="list-style-type: decimal">
<li>
<strong>tl</strong>: The length of the metabolic labeling feed. This
can be in any units (the simulated data puts it in terms of minutes),
but <strong>if no s<sup>4</sup>U was fed to a sample, tl must be 0 for
that sample</strong>. -s<sup>4</sup>U controls play an important role in
all of the statistical models, so this is crucial to remember. While not
technically necessary to run bakR, we always highly suggest including
these controls.</li>
<li>
<strong>Exp_ID</strong>: Stands for “Experimental ID” and is a
numerical ID that denotes which samples are from the reference condition
(Exp_ID = 1) and which experimental samples belong to the same
“treatment” (Exp_IDs &gt; 1). You can have more than one experimental
sample, but all comparisons will only be done with respect to one
reference sample.</li>
</ol>
<p>The metadf data frame must also have row names corresponding to the
sample name that the <strong>tl</strong> and <strong>Exp_ID</strong>
entries describe, as the sample name appears in the cB data frame.</p>
<p>Once you have these two data frames correctly constructed, you can
create a bakRData object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"cB_small"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"metadf"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRData.html">bakRData</a></span><span class="op">(</span><span class="va">cB_small</span>, <span class="va">metadf</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-fitting-the-efficient-model">Step 2: Fitting the Efficient Model<a class="anchor" aria-label="anchor" href="#step-2-fitting-the-efficient-model"></a>
</h2>
<p>bakR contains three different implementations of a statistical model
of NR-seq data. You must always first run the most efficient
implementation. The other two implementations can then be run using the
output of this implementation. Below I will show how to run the
efficient implementation using simulated data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_bakRData.html">Simulate_bakRData</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 500 features, 2 experimental conditions</span></span>
<span>  <span class="co"># and 3 replicates for each experimental condition</span></span>
<span>  <span class="co"># See ?Simulate_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Extract simulated bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span></span>
<span></span>
<span><span class="co"># Extract simualted ground truths</span></span>
<span><span class="va">sim_truth</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">sim_list</span></span>
<span></span>
<span><span class="co"># Run the efficient model</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew    pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.050<span style="text-decoration: underline;">2</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     1     3 0.050<span style="text-decoration: underline;">1</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     1 0.050<span style="text-decoration: underline;">1</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     2     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     2     3 0.050<span style="text-decoration: underline;">3</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>bakRFit() is used here as a wrapper for two functions in bakR:
cBprocess() and fast_analysis(). For more details on what these
functions do, run <code><a href="../reference/cBprocess.html">?cBprocess</a></code> or <code>fast_analysis</code>.
Alternatively, see the more highly detailed version of this vignette for
additional details.</p>
<p>While the fast implementation was running, it outputted a message
regarding the estimated pnews and pold. The pnews are the estimated
mutation rates of reads from new RNAs (new meaning RNAs synthesized
after the start of s<sup>4</sup>U labeling) in each sample (muts =
Exp_ID, and reps = a numerical replicate ID that corresponds to the
order replicates appear in the cB), and polds are the estimates for the
background mutation rate used in all analyses. For more details on how
bakR estimates these mutation rates and alternative estimation
strategies implemented in bakR, see the longer form version of this
vignette as well as the vignette on troubleshooting analyses.</p>
<p>To run the heavier, more highly powered models, just rerun bakRFit()
on the Fit object, but with either the StanFit or HybridFit parameters
set to true.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run Hybrid model (This might take several minutes to run)</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">Fit</span>, HybridFit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run Full model (This might take ~10-30 minutes to run)</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">Fit</span>, StanFit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The Fit objects contain lists pertaining to the fits of each of the
models. The possible contents include:</p>
<ol style="list-style-type: decimal">
<li>Fast_Fit: Result of the initial fitting of the bakRData object. To
learn more about its contents, see ?fast_analysis()</li>
<li>Data_lists: Processed data that can be passed to the statistical
models</li>
<li>Hybrid_Fit: Result of running bakRFit() on the Fit object with
HybridFit = TRUE. To learn more about its contents, see ?TL_stan().</li>
<li>Stan_Fit: Result of running bakRFit() on the Fit object with StanFit
= TRUE. Its general contents are identical to the Hybrid_Fit, even
though the models are different.</li>
</ol>
</div>
<div class="section level2">
<h2 id="step-3-visualizing-the-results">Step 3: Visualizing the Results<a class="anchor" aria-label="anchor" href="#step-3-visualizing-the-results"></a>
</h2>
<p>bakR provides a variety of easy to use functions for beginning to
investigate your data. The visualizations are particularly aimed at
revealing trends in RNA stabilization or destabilization. These include
MA plots:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## MA Plot with Fast Fit</span></span>
<span><span class="fu">bakR</span><span class="fu">::</span><span class="fu"><a href="../reference/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">Fit</span>, Model <span class="op">=</span> <span class="st">"MLE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Quickstart_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Volcano plots:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Volcano Plot with Fast Fit; significance assessed relative to an FDR control of 0.05</span></span>
<span><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Quickstart_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>and PCA plots:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2D PCA plot with replicate fraction news</span></span>
<span>  <span class="co"># The equivalent function prior to version 1.0.0 is FnPCA, now deprecated in </span></span>
<span>  <span class="co"># favor of FnPCA2.</span></span>
<span><span class="fu"><a href="../reference/FnPCA2.html">FnPCA2</a></span><span class="op">(</span><span class="va">Fit</span>, Model <span class="op">=</span> <span class="st">"MLE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bakR-Quickstart_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="step-4-where-to-go-from-here">Step 4: Where to go from here<a class="anchor" aria-label="anchor" href="#step-4-where-to-go-from-here"></a>
</h2>
<p>This vignette provides the minimal amount of information to get up
and running with bakR. If you would like a more thorough discussion of
each step of this process, check out the long form version of this
vignette (“Differential kinetic analysis with bakR”). In addition, there
are a number of other vignettes that cover various topics not discussed
in these intro vignettes:</p>
<ol style="list-style-type: decimal">
<li>Differential synthesis rate analysis</li>
<li>Correcting for loss of s<sup>4</sup>U RNA during library prepartion
(a.k.a. dropout correction)</li>
<li>Troubleshooting analyses of NR-seq datasets</li>
<li>Mechanistic dissection of differential expression</li>
<li>Running bakR with fraction new estimate input (e.g., from
GRAND-SLAM)</li>
</ol>
<p>Note, all but the “Differential synthesis rate analysis” vignette are
only fully compatible with version 1.0.0 of bakR. Update bakR as
necessary</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
