<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bakR">
<title>Correcting for dropout • bakR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Correcting for dropout">
<meta property="og:description" content="bakR">
<meta property="og:image" content="https://simonlabcode.github.io/bakR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a>
    <a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a>
    <a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a>
    <a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a>
    <a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a>
    <a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Correcting for dropout</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/vignettes/Dropout.Rmd" class="external-link"><code>vignettes/Dropout.Rmd</code></a></small>
      <div class="d-none name"><code>Dropout.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Dropout is the phenomenon by which metabolically labeled RNA or
      sequencing reads derived from such RNA is lost either during
      library preparation or during computational processing of the raw
      sequencing data. Several labs have independently identified and
      discussed the prevalence of dropout in NR-seq and other metabolic
      labeling RNA-seq experiments. Recent work suggests that there are
      three potential causes for this: 1) Loss of labeled RNA on plastic
      surfaces during RNA extraction, 2) RT falloff due to modifications
      of the metabolic labeling made by some NR-seq chemistries, and 3)
      loss of reads with many NR-induced mutations due to poor read
      alignment. While 3) is best remedied by improved alignment
      strategies (discussed in <a href="https://www.biorxiv.org/content/10.1101/2023.05.24.542133v1" class="external-link">Zimmer
      et al.</a> and <a href="https://www.biorxiv.org/content/10.1101/2023.04.21.537786v1" class="external-link">Berg
      et al.</a>), modified fastq processing will not address the other
      causes. Improved handling of RNA can help, but what if you already
      have data that you suspect may be biased by dropout? Version 1.0.0
      of bakR introduced a strategy to correct for dropout induced
      biases. In this vignette, I will show how to correct for dropout
      with bakR. Following the demonstrations, I will also include a
      more thorough discussion of dropout, bakR’s dropout correction
      strategy, and how it differs from a <a href="https://www.biorxiv.org/content/10.1101/2023.04.21.537786v1" class="external-link">strategy
      recently developed</a> by the Erhard lab.</p>
    </div>
    
<div class="section level2">
<h2 id="necessary-setup">Necessary Setup<a class="anchor" aria-label="anchor" href="#necessary-setup"></a>
</h2>
<p>bakR version 1.0.0 or later is necessary to run this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="a-brief-tutorial-on-dropout-correction-in-bakr">A Brief Tutorial on Dropout Correction in bakR<a class="anchor" aria-label="anchor" href="#a-brief-tutorial-on-dropout-correction-in-bakr"></a>
</h2>
<p>bakR has a new simulation strategy as of version 1.0.0 that allows
for the accurate simulation of dropout:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_relative_bakRData.html">Simulate_relative_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>, depth <span class="op">=</span> <span class="fl">1000000</span>, nreps <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                       p_do <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 500 features, 500,000 reads, 2 experimental conditions</span></span>
<span>  <span class="co"># and 2 replicates for each experimental condition.</span></span>
<span>  <span class="co"># 40% dropout is simulated.</span></span>
<span>  <span class="co"># See ?Simulate_relative_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Run the efficient model</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew    pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.049<span style="text-decoration: underline;">7</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2     1 0.049<span style="text-decoration: underline;">9</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     2 0.049<span style="text-decoration: underline;">7</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>Dropout will cause bakR to underestimate the true fraction new. In
addition, low fraction new features will receive more read counts than
they would have if dropout had not existed (and vice versa for high
fraction new features). Both of these biases can be corrected for with a
single line of code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Correct dropout-induced biases</span></span>
<span><span class="va">Fit_c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CorrectDropout.html">CorrectDropout</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimated rates of dropout are:</span></span>
<span><span class="co">#&gt;   Exp_ID Replicate       pdo</span></span>
<span><span class="co">#&gt; 1      1         1 0.4013715</span></span>
<span><span class="co">#&gt; 2      1         2 0.3841179</span></span>
<span><span class="co">#&gt; 3      2         1 0.2627656</span></span>
<span><span class="co">#&gt; 4      2         2 0.3242958</span></span>
<span><span class="co">#&gt; Mapping sample name to sample characteristics</span></span>
<span><span class="co">#&gt; Filtering out low coverage features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># You can also overwite the existing bakRFit object.</span></span>
<span>  <span class="co"># I am creating a separate bakRFit object to make comparisons later in this vignette.</span></span></code></pre></div>
<p>Dropout will lead to a correlation between the fraction new and the
difference between +s<sup>4</sup>U and -s<sup>4</sup>U read counts. This
trend and the model fit can be visualized as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Correct dropout-induced biases</span></span>
<span><span class="va">Vis_DO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VisualizeDropout.html">VisualizeDropout</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; Estimated rates of dropout are:</span></span>
<span><span class="co">#&gt;   Exp_ID Replicate       pdo</span></span>
<span><span class="co">#&gt; 1      1         1 0.4013715</span></span>
<span><span class="co">#&gt; 2      1         2 0.3841179</span></span>
<span><span class="co">#&gt; 3      2         1 0.2627656</span></span>
<span><span class="co">#&gt; 4      2         2 0.3242958</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Visualize dropout for 1st replicate of reference condition</span></span>
<span><span class="va">Vis_DO</span><span class="op">$</span><span class="va">ExpID_1_Rep_1</span></span></code></pre></div>
<p><img src="Dropout_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can assess the impact of dropout correction by comparing the known
simulated truth to the fraction new estimates, before and after dropout
correction.</p>
<p>Before correction:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Extract simualted ground truths</span></span>
<span><span class="va">sim_truth</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">sim_list</span></span>
<span></span>
<span><span class="co"># Features that made it past filtering</span></span>
<span><span class="va">XFs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">XF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulated logit(fraction news) from features making it past filtering</span></span>
<span><span class="va">true_fn</span> <span class="op">&lt;-</span> <span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Logit_fn</span><span class="op">[</span><span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Feature_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">XFs</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Estimated logit(fraction news)</span></span>
<span><span class="va">est_fn</span> <span class="op">&lt;-</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">logit_fn</span></span>
<span></span>
<span><span class="co"># Compare estimate to truth</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">true_fn</span>, <span class="va">est_fn</span>, xlab <span class="op">=</span> <span class="st">"True logit(fn)"</span>, ylab <span class="op">=</span> <span class="st">"Estimated logit(fn)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Dropout_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>After correction:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Features that made it past filtering</span></span>
<span><span class="va">XFs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Fit_c</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">XF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulated logit(fraction news) from features making it past filtering</span></span>
<span><span class="va">true_fn</span> <span class="op">&lt;-</span> <span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Logit_fn</span><span class="op">[</span><span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Feature_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">XFs</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Estimated logit(fraction news)</span></span>
<span><span class="va">est_fn</span> <span class="op">&lt;-</span> <span class="va">Fit_c</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">logit_fn</span></span>
<span></span>
<span><span class="co"># Compare estimate to truth</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">true_fn</span>, <span class="va">est_fn</span>, xlab <span class="op">=</span> <span class="st">"True logit(fn)"</span>, ylab <span class="op">=</span> <span class="st">"Estimated logit(fn)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Dropout_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="rna-dropout-and-bakrs-correction-strategy">RNA dropout and bakR’s correction strategy<a class="anchor" aria-label="anchor" href="#rna-dropout-and-bakrs-correction-strategy"></a>
</h2>
<p>As discussed in the Abstract for this vignette, dropout refers to the
loss of metabolically labeled RNA, or reads from said RNA. Dropout
correction in the context of bakR is specifically designed to address
biases due to the loss of metabolically labeled RNA during library
preparation.</p>
<p>bakR’s strategy for correcting dropout involves formalizing a model
for dropout, using that model to infer a parametric relationship between
the biased fraction new estimate and the ratio of +s<sup>4</sup>U to
-s<sup>4</sup>U read counts for a feature, and then fitting that model
to data with nonlinear least squares. bakR’s model of dropout makes
several simplifying assumptions to ensure tractability of parameter
estimation while still capturing important features of the process:</p>
<ol style="list-style-type: decimal">
<li>The average number of reads that will come from an RNA feature is
related to the fraction of sequenced RNA molecules derived from that
feature.</li>
<li>All labeled RNA is equally likely to get lost during library
preparation, with the probability of losing a molecule of labeled RNA
referred to as pdo.</li>
</ol>
<p>These assumptions lead to the following expressions for the expected
number of sequencing reads coming from a feature (index i) in
+s<sup>4</sup>U and -s<sup>4</sup>U NR-seq data:</p>
<p><span class="math display">\[
\begin{align}
\text{E[reads from feature i in +s}^{\text{4}}\text{U sample]} &amp;=
\frac{\text{number of molecules from feature i}}{\text{total number of
molecules in +s}^{\text{4}}\text{U
sample}}*\text{R}_{\text{+s}^{\text{4}}\text{U}} \\
                                              &amp;=
\frac{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 - \text{pdo}) +
\text{n}_{\text{i}}*(1 - \theta_{\text{i}})
}{\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}*\theta_{\text{j}}*(1
- \text{pdo}) + \text{n}_{\text{j}}*(1 - \theta_{\text{j}})
}*\text{R}_{\text{+s}^{\text{4}}\text{U}} \\
                                              \\
                                              \\
\text{E[reads from feature i in -s}^{\text{4}}\text{U sample]} &amp;=
\frac{\text{number of molecules from feature i}}{\text{total number of
molecules in -s}^{\text{4}}\text{U
sample}}*\text{R}_{\text{-s}^{\text{4}}\text{U}} \\
                                              &amp;=
\frac{\text{n}_{\text{i}}}{\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}}*\text{R}_{\text{-s}^{\text{4}}\text{U}}
\end{align}                                  
\]</span></p>
<p>where the parameters are defined as follows:</p>
<p><span class="math display">\[
\begin{align}
\text{n}_{\text{i}} &amp;= \text{Number of molecules from feature i} \\
\theta_{\text{i}} &amp;= \text{True fraction new for feature i} \\
\text{pdo} &amp;= \text{Probability that a labeled RNA molecule is
preferentially lost during library prep} \\
\text{NF} &amp;= \text{Total number of RNA features that contribute
sequencable molecules} \\
\text{R} &amp;= \text{Total number of sequencing reads.}
\end{align}
\]</span></p>
<p>Defining “dropout” to be the ratio of the RPM normalized expected
read counts yields the following relationship between “dropout”, the
fraction new, and the rate of dropout:</p>
<p><span class="math display">\[
\begin{align}
\text{Dropout} &amp;= \frac{\text{E[reads from feature i in
+s}^{\text{4}}\text{U
sample]/}\text{R}_{\text{+s}^{\text{4}}\text{U}}}{\text{E[reads from
feature i in -s}^{\text{4}}\text{U
sample]/}\text{R}_{\text{-s}^{\text{4}}\text{U}}} \\
\\
               &amp;= \frac{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 -
\text{pdo}) + \text{n}_{\text{i}}*(1 - \theta_{\text{i}})
}{\text{n}_{\text{i}}}*\frac{\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}}{
\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}*\theta_{\text{j}}*(1 -
\text{pdo}) + \text{n}_{\text{j}}*(1 - \theta_{\text{j}})  } \\
               \\
               &amp;= [\theta_{\text{i}}*(1 - \text{pdo}) + (1 -
\theta_{\text{i}})]*\text{scale} \\
               &amp;= [1 - \theta_{\text{i}}*\text{pdo}]*\text{scale}
\end{align}
\]</span></p>
<p>where <span class="math inline">\(\text{scale}\)</span> is a constant
scale factor that represents the factor difference in the number of
sequencable molecules with and without dropout.</p>
<p>Currently, the relationship between the quantification of dropout and
<span class="math inline">\(\text{pdo}\)</span> includes <span class="math inline">\(\theta_{\text{i}}\)</span>, which is the true,
unbiased fraction new. Unfortunately, the presence of dropout means that
we do not have access to this quantify. Rather, we estimate a dropout
biased fraction of sequencing reads that are new (<span class="math inline">\(\theta_{\text{i}}^{\text{do}}\)</span>), which on
average will represent an underestimation of <span class="math inline">\(\theta_{\text{i}}\)</span>. Therefore, we need to
relate the quantity we estimate (<span class="math inline">\(\theta_{\text{i}}^{\text{do}}\)</span>) and the
parameter we wish to estimate (<span class="math inline">\(\text{pdo}\)</span>) to <span class="math inline">\(\theta_{\text{i}}\)</span>. In the context of this
model, such a relationship can be derived as follows:</p>
<p><span class="math display">\[
\begin{align}
\theta_{\text{i}}^{\text{do}} &amp;= \frac{\text{number of new
sequencable molecules}}{\text{total number of sequencable molecules}} \\
\theta_{\text{i}}^{\text{do}}                               &amp;=
\frac{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 - \text{pdo})
}{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 - \text{pdo}) +
\text{n}_{\text{i}}*(1-\theta_{\text{i}})} \\
\theta_{\text{i}}^{\text{do}}                              &amp;=
\frac{\theta_{\text{i}}*(1 - \text{pdo}) }{\theta_{\text{i}}*(1 -
\text{pdo}) + (1-\theta_{\text{i}})} \\
\theta_{\text{i}}^{\text{do}}*[1 - \theta_{\text{i}}*\text{pdo}] &amp;=
\theta_{\text{i}}*(1 - \text{pdo}) \\
\theta_{\text{i}}^{\text{do}} &amp;= \theta_{\text{i}}*(1 - \text{pdo})
+ \theta_{\text{i}}*\theta_{\text{i}}^{\text{do}}*\text{pdo} \\
\frac{\theta_{\text{i}}^{\text{do}}}{(1 - \text{pdo}) +
\theta_{\text{i}}^{\text{do}}*\text{pdo}} &amp;= \theta_{\text{i}}
\end{align}
\]</span></p>
<p>We can then use this relationship to substite <span class="math inline">\(\theta_{\text{i}}\)</span> for a function of <span class="math inline">\(\theta_{\text{i}}^{\text{do}}\)</span> and <span class="math inline">\(\text{pdo}\)</span> in our dropout quantification
equation:</p>
<p><span class="math display">\[
\begin{align}
\text{Dropout} &amp;= [1 - \theta_{\text{i}}*\text{pdo}]*\text{scale}\\
\text{Dropout} &amp;= \text{scale} -
\frac{\text{scale}*\text{pdo}*\theta_{\text{i}}^{\text{do}}}{(1 -
\text{pdo}) + \theta_{\text{i}}^{\text{do}}*\text{pdo}}
\end{align}
\]</span></p>
<p>bakR’s <code>CorrectDropout</code> function fits this predicted
relationship between the ratio of +s<sup>4</sup>U to -s<sup>4</sup>U
reads (<span class="math inline">\(\text{Dropout}\)</span>) and the
uncorrected fraction new estimates (<span class="math inline">\(\theta_{\text{i}}^{\text{do}}\)</span>) to infer
<span class="math inline">\(pdo\)</span>. Corrected fraction new
estimates can then be inferred from the relationship between <span class="math inline">\(\theta_{\text{i}}^{\text{do}}\)</span>, <span class="math inline">\(\text{pdo}\)</span>, and <span class="math inline">\(\theta_{\text{i}}\)</span>. Finally, bakR
redistributes read counts according to what the estimated relative
proportions of each feature would have been if no dropout had existed.
The key insight is that:</p>
<p><span class="math display">\[
\begin{align}
\frac{\text{E[number of reads without dropout]}}{\text{E[number of reads
with dropout]}} &amp;= \text{Dropout}\\
&amp;= \frac{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 - \text{pdo}) +
\text{n}_{\text{i}}*(1 - \theta_{\text{i}})
}{\text{n}_{\text{i}}}*\frac{\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}}{
\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}*\theta_{\text{j}}*(1 -
\text{pdo}) + \text{n}_{\text{j}}*(1 - \theta_{\text{j}})  } \\
&amp;= \frac{\theta_{\text{i}}*(1 - \text{pdo}) + (1 -
\theta_{\text{i}})
}{1}*\frac{\text{N}}{\text{N}*\theta_{\text{G}}*(1-\text{pdo}) +
\text{N}*(1 - \theta_{\text{G}})}\\
&amp;= \frac{\theta_{\text{i}}*(1 - \text{pdo}) + (1 -
\theta_{\text{i}}) }{1}*\frac{1}{\theta_{\text{G}}*(1 - \text{pdo}) + (1
- \theta_{\text{G}}) } \\
\\
&amp;= \frac{\text{fraction of feature i molecules remaining after
dropout}}{\text{fraction of total molecules remaining after dropout}} \\
\end{align}
\]</span></p>
<p>Getting from the 2nd line to the third line involved a bit of
algebraic trickery (multiplying by 1) and defining the “global fraction
new” (<span class="math inline">\(\theta_{\text{G}}\)</span>), which is
the fraction of all sequenced molecules that are new (where <span class="math inline">\(\text{N}\)</span> is the total number of sequenced
molecules if none are lost due to dropout). <span class="math inline">\(\theta_{\text{G}}\)</span> can be calculated as a
weighted average of dropout biased fraction news for each feature,
weighted by the uncorrected number of reads each feature has, and then
dropout correcting:</p>
<p><span class="math display">\[
\begin{align}
\theta_{\text{G}}^{\text{do}} &amp;=
\frac{\sum_{\text{j=1}}^{\text{NF}}\theta_{\text{j}}^{\text{do}}*\text{nreads}_{\text{j}}
}{\sum_{\text{j=1}}^{\text{NF}}\text{nreads}_{\text{j}}} \\
\theta_{\text{G}} &amp;= \frac{\theta_{\text{G}}^{\text{do}}}{(1 -
\text{pdo}) + \theta_{\text{G}}^{\text{do}}*\text{pdo}}
\end{align}
\]</span></p>
<p>Thus, the dropout corrected read counts can be obtained as
follows:</p>
<p><span class="math display">\[
\begin{align}
\text{Corrected read count for feature i} &amp;=
\text{nreads}_{\text{i}}*\frac{\theta_{\text{G}}*(1 - \text{pdo}) + (1 -
\theta_{\text{G}})}{\theta_{\text{i}}*(1 - \text{pdo}) + (1 -
\theta_{\text{i}})}
\end{align}
\]</span></p>
</div>
<div class="section level2">
<h2 id="differences-between-bakr-and-grandrs-strategies">Differences between bakR and grandR’s strategies<a class="anchor" aria-label="anchor" href="#differences-between-bakr-and-grandrs-strategies"></a>
</h2>
<p>Florian Erhard’s group was the first to implement a strategy for
dropout correction in an NR-seq analysis tool (their R package <a href="https://github.com/erhard-lab/grandR" class="external-link">grandR</a>). The strategy
used by grandR is discussed <a href="https://pubmed.ncbi.nlm.nih.gov/38381903/" class="external-link">here</a>. In short, a
factor <span class="math inline">\(f\)</span> is found such that if the
number of new reads (inferred as the fraction new * total reads) is
multiplied by <span class="math inline">\(f\)</span>, then the Spearman
correlation between the fraction new and the difference in -s4U and +s4U
is 0. The dropout rate is then calculated as:</p>
<p><span class="math display">\[
\begin{align}
\text{pdo} &amp;= \frac{f}{f + 1}
\end{align}
\]</span></p>
<p>We note that this definition is not derived from an explicit model.
The adjusted fraction new is then calculated as:</p>
<p><span class="math display">\[
\begin{align}
\text{corrected fn} = \frac{f*(\text{number of reads from new
RNA})}{f*(\text{number of reads from new RNA}) + (\text{number of reads
from old RNA})}
\end{align}
\]</span></p>
<p><strong>NOTE: previous versions of this vignette included some
inaccuracies about the Erhard lab’s method. These have been corrected
and we apologize for the mistakes. While the explanation of their method
here is not identical to that provided in the methods section of their
paper, our explanation follows from grandR’s source code.</strong> A bit
of algebra shows that this relationship between the inferred <span class="math inline">\(\text{pdo}\)</span> and the corrected fraction new
is identical to that derived from the model above:</p>
<p><span class="math display">\[
\begin{align}
\theta_{\text{i}} &amp;= \frac{f*(\text{number of reads from new
RNA})}{f*(\text{number of reads from new RNA}) + (\text{number of reads
from old RNA})} \\
&amp;=
\frac{f*\theta_{\text{i}}^{\text{do}}*\text{nreads}_{\text{i}}}{f*\theta_{\text{i}}^{\text{do}}*\text{nreads}_{\text{i}}
+ (1 - \theta_{\text{i}}^{\text{do}})*\text{nreads}_{\text{i}}} \\
&amp;= \frac{\frac{\theta_{\text{i}}^{\text{do}}}{(1 -
\text{pdo})}}{\frac{\theta_{\text{i}}^{\text{do}}}{(1 - \text{pdo})} +
(1 - \theta_{\text{i}}^{\text{do}})} \\
&amp;= \frac{\theta_{\text{i}}^{\text{do}}}{(1 - \text{pdo}) +
\theta_{\text{i}}^{\text{do}}*\text{pdo}}
\end{align}
\]</span></p>
<p>Thus, grandR is implicitly making assumptions similar to those laid
out in the model above that is used by bakR.</p>
<p>Notable differences between bakR and grandR’s dropout correction
approaches are:</p>
<ol style="list-style-type: decimal">
<li>bakR defines an explicit model of dropout and derives a strategy to
estimate the dropout rate from data. grandR defines an ad hoc
relationship between a fraction new vs. dropout correlation coefficient
and the dropout rate.</li>
<li>When correcting for read counts, grandR seems to use a different
correction strategy than is derived from the model above. New read
counts are multiplied by the inferred factor <span class="math inline">\(f\)</span> discussed earlier. bakR on the other
hand redistributes read counts, thus preserving the total library
size.</li>
<li>bakR fits an explicit parametric model derived from assumptions made
either explicitly or implicitly by both grandR and bakR. This allows
users to assess the quality of the model fit and thus the likelihood
that these assumptions are valid on any given dataset.</li>
</ol>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="a1-algebraic-aside">A1: Algebraic aside<a class="anchor" aria-label="anchor" href="#a1-algebraic-aside"></a>
</h3>
<p>One potentially non-obvious step in the derivations above is when
determining how to correct read counts. In that section, I had the
following set of relationships:</p>
<p><span class="math display">\[
\begin{align}
\frac{\text{E[number of reads without dropout]}}{\text{E[number of reads
with dropout]}} &amp;= \text{Dropout}\\
&amp;= \frac{\text{n}_{\text{i}}*\theta_{\text{i}}*(1 - \text{pdo}) +
\text{n}_{\text{i}}*(1 - \theta_{\text{i}})
}{\text{n}_{\text{i}}}*\frac{\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}}{
\sum_{\text{j}=1}^{\text{NF}}\text{n}_{\text{j}}*\theta_{\text{j}}*(1 -
\text{pdo}) + \text{n}_{\text{j}}*(1 - \theta_{\text{j}})  } \\
&amp;= \frac{\theta_{\text{i}}*(1 - \text{pdo}) + (1 -
\theta_{\text{i}})
}{1}*\frac{\text{N}}{\text{N}*\theta_{\text{G}}*(1-\text{pdo}) +
\text{N}*(1 - \theta_{\text{G}})} \\
&amp;=\text{...}
\end{align}
\]</span></p>
<p>where:</p>
<p><span class="math display">\[
\begin{align}
\text{N} &amp;= \sum_{\text{j=1}}^{\text{NF}}{\text{n}_{\text{j}}} \\
\theta_{\text{G}} &amp;=
\frac{\sum_{\text{j=1}}^{\text{NF}}\text{n}_{\text{j}}*\theta_{\text{i}}}{\sum_{\text{j=1}}^{\text{NF}}\text{n}_{\text{j}}}
\end{align}
\]</span></p>
<p>The algebraic trick to get from line 2 to line 3 is to multiply by 1
(or rather <span class="math inline">\(\text{N}\)</span>/<span class="math inline">\(\text{N}\)</span>):</p>
<p><span class="math display">\[
\begin{align}
\sum_{\text{j=1}}^{\text{NF}}\text{n}_{\text{i}}*\theta_{\text{j}}
&amp;=
\sum_{\text{j=1}}^{\text{NF}}{\text{n}_{\text{j}}}*\frac{\sum_{\text{j=1}}^{\text{NF}}\text{n}_{\text{i}}*\theta_{\text{j}}}{\sum_{\text{j=1}}^{\text{NF}}{\text{n}_{\text{j}}}}
\\
&amp;= \text{N}*\theta_{\text{G}}
\end{align}
\]</span></p>
</div>
<div class="section level3">
<h3 id="a2-u-content-dependence">A2: U-content dependence<a class="anchor" aria-label="anchor" href="#a2-u-content-dependence"></a>
</h3>
<p>Given hypotheses about how dropout arises, you may expect the extent
of dropout to be correlated with the U-content of the RNA feature. In
fact, this does seem to be the case in real datasets. bakR currently
does not attempt to account for this correlation for a few reasons:</p>
<ol style="list-style-type: decimal">
<li>U-content is not typically correlated with estimated fraction new.
Thus, the U-content dropout biases are often evenly represented
throughout the fraction new vs. dropout trend. This means that their
influence is to add scatter to the data while not perturbing the mean of
the trend.</li>
<li>A rigorous empirical relationship between U-content and dropout is
difficult to derive from first principles given existing data. An ad hoc
correction factor can be added, but so far we have found it to have
minimal impact on dropout rate estimation.</li>
</ol>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
