<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bakR">
<title>Going beyond the defaults with bakR • bakR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Going beyond the defaults with bakR">
<meta property="og:description" content="bakR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/Further-Analyses.html">Going beyond the defaults with bakR</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Going beyond the defaults with bakR</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/vignettes/Further-Analyses.Rmd" class="external-link"><code>vignettes/Further-Analyses.Rmd</code></a></small>
      <div class="d-none name"><code>Further-Analyses.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Welcome! This vignette discusses an exciting hodge-podge of
additional analyses that can be done with bakR. The assumption is that
at this point you have worked through the Getting-Started vignette. If
that is not the case, I would highly suggest starting there and then
coming back here once you’re familiar with the standard use cases of
bakR. The topics to be addressed are as follows:</p>
<ol style="list-style-type: decimal">
<li>Performing differential synthesis analysis with bakR (and DESeq2, as
you will see)</li>
<li>Assessing the mechanisms of differential expression in non-steady
state systems</li>
<li>TBD (Keep your eyes for announcements about updates to this vignette
with more to come!)</li>
</ol>
<p>These are the packages you are going to have to install and load in
order to do everything presented in this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Packages that are NOT automatically installed when bakR is installed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:bakR':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plotMA</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Packages which are installed when bakR is installed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     count</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'magrittr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     subtract</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set the seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-synthesis-analysis-with-bakr-deseq2">Differential synthesis analysis with bakR + DESeq2<a class="anchor" aria-label="anchor" href="#differential-synthesis-analysis-with-bakr-deseq2"></a>
</h2>
<p>I claim that bakR is a tool for performing differential kinetic
analysis, but if you are coming from the Getting-Started vignette then
you might think that this is a misnomer. All that I showed you how to do
in that vignette was differential <em>stability</em> analysis.
Full-fledged differential kinetic analysis means performing differential
<em>synthesis</em> analysis as well, so let’s talk about how you can do
that with bakR.</p>
<p>As the title of this section suggests though, bakR won’t be able to
accomplish this task alone. The reason for this is that assessing
changes in synthesis means assessing changes in RNA stability and RNA
expression. Therefore, we need to perform differential expression
analysis in conjunction with differential stability analysis. In theory,
I could have implemented my own differential expression analysis in
bakR, but with so many popular software tools currently in existence to
perform that task, it seemed pointlessly redundant. It also gives you
the freedom to use whatever differential expression analysis tool that
you are used to using, and to perform whatever kind of DE analysis that
you like independent of bakR.</p>
<p>To show you how to perform differential synthesis analysis, I will be
using DESeq2 for the purpose of differential expression analysis. Why?
Because I like DESeq2. It’s a good model, it’s a good piece of software,
and it was an early source of inspiration for me in my PhD. Also, you’ll
see that some of the output of bakR is specifically designed with DESeq2
in mind, so it will facilitate the process of performing differential
expression analysis a bit.</p>
<p>First things first, let’s simulate some data! We’ll simulate 1000
genes, with 200 observing differential synthesis and no changes in
stability. Let’s also run the efficient implementation while we are at
it.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_bakRData.html">Simulate_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>,</span>
<span>                         num_kd_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                         num_ks_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 500 features, 2 experimental conditions</span></span>
<span>  <span class="co"># and 3 replicates for each experimental condition</span></span>
<span>  <span class="co"># See ?Simulate_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Extract simulated bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span></span>
<span></span>
<span><span class="co"># Extract simualted ground truths</span></span>
<span><span class="va">sim_truth</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">sim_list</span></span>
<span></span>
<span><span class="co">## Run the efficient model</span></span>
<span><span class="co"># We'll tell it what the mutation rates are just for efficiency's sake</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bakRData</span>, pnew <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.05</span>, times <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, pold <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>The second step is to perform differential expression analysis. As I
alluded to above, bakR makes this a bit easier, and the way it does that
is by providing a count matrix formatted just how DESeq2 likes it!
DESeq2 also requires one more input, and that is a dataframe that
provides information about any covariates (things that distinguish
samples). Here is all the code to produce those necessary inputs</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Get the count matrix from bakR</span></span>
<span><span class="va">Counts</span> <span class="op">&lt;-</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Data_lists</span><span class="op">$</span><span class="va">Count_Matrix</span></span>
<span></span>
<span><span class="co"># Experimental conditions for each sample</span></span>
<span><span class="co"># There are 6 s4U treated samples (3 replicates of each condition)</span></span>
<span><span class="co"># In addition, there are 2 -s4U control samples (1 for each condition)</span></span>
<span></span>
<span><span class="co">## s4U conditions</span></span>
<span><span class="co"># 1st three samples are reference (ref) samples</span></span>
<span><span class="co"># Next three samples are experimental (exp) samples</span></span>
<span><span class="va">conditions_s4U</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## -s4U control conditions</span></span>
<span><span class="co"># 1st sample is reference, next is experimental</span></span>
<span><span class="va">conditions_ctl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined s4U and -s4U control conditions</span></span>
<span><span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">conditions_s4U</span>, <span class="va">conditions_ctl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make the colData input for DESeq2</span></span>
<span><span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>conditions <span class="op">=</span> <span class="va">conditions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">colData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the colData object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">colData</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">1</th>
<th align="left">2</th>
<th align="left">3</th>
<th align="left">4</th>
<th align="left">5</th>
<th align="left">6</th>
<th align="left">7</th>
<th align="left">8</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">conditions</td>
<td align="left">ref</td>
<td align="left">ref</td>
<td align="left">ref</td>
<td align="left">exp</td>
<td align="left">exp</td>
<td align="left">exp</td>
<td align="left">ref</td>
<td align="left">exp</td>
</tr></tbody>
</table>
<p>Once you make sure that you have correctly mapped samples to
experimental conditions in the colData object, you can create that sweet
sweet DESeqDataObject</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">Counts</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">colData</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span><span class="va">conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span></code></pre></div>
<p>I invite you to read the DESeq2 documentation if you need help
understanding the call to DESeqDataSetFromMatrix. In short though, the
first entry is our count matrix, the second is our colData object, and
the final entry is the design matrix. Limma’s documentation (another
differential expression analysis software) has a great introduction to
design matrices, but this one is simple. It means that we want to group
samples together that have the same value for the conditions factor in
the colData object. So all “ref” samples will be grouped together and
all “exp” samples will be grouped together. DESeq2 will then compare
different groups, which means that it will compare “exp” and “ref”
samples in this case.</p>
<p>Now we can fit DESeq2 and extract the differential expression
analysis results:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ddso</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Extract results of experimental vs. reference comparison</span></span>
<span><span class="va">reso</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">ddso</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"conditions"</span>, <span class="st">"exp"</span>, <span class="st">"ref"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the column names of the reso object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">reso</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "baseMean"       "log2FoldChange" "lfcSE"          "stat"          </span></span>
<span><span class="co">#&gt; [5] "pvalue"         "padj"</span></span></code></pre></div>
<p>For our sake, we are going to use two columns in the reso object:
log2FoldChange and lfcSE. Both of these column names are pretty
self-explanatory, but for completeness sake, log2FoldChange is the log
base-2 fold change in RNA expression, and lfcSE is the standard error in
the log2FoldChange.</p>
<p>With that, we have all we need to perform differential synthesis
analysis! But how you may ask? Well, if our population of cells that we
performed RNA-seq on was at steady-state (i.e., the cells were not
actively responding to some perturbation while treating with with a
metabolic label), then the following relationship holds between <span class="math inline">\(ksyn\)</span> (RNA synthesis rate) <span class="math inline">\(kdeg\)</span> (RNA degradation rate) and <span class="math inline">\([RNA]\)</span> (RNA concentration):</p>
<p><span class="math display">\[
[RNA] = \frac{ksyn}{kdeg}
\]</span></p>
<p>Therefore, the log2 fold change in <span class="math inline">\([RNA]\)</span> (<span class="math inline">\(L2FC(RNA)\)</span>) is: <span class="math display">\[
\begin{align}
L2FC(RNA) &amp;= log_2(\frac{[RNA_{exp}]}{[RNA_{ref}]}) \\
          &amp;=
log_2(\frac{ksyn_{exp}*kdeg_{ref}}{ksyn_{ref}*kdeg_{exp}}) \\
          &amp;= log_2(\frac{ksyn_{exp}}{ksyn_{ref}}) -
log_2(\frac{kdeg_{exp}}{kdeg_{ref}}) \\
          &amp;= L2FC(ksyn) - L2FC(kdeg)
\end{align}
\]</span> Therefore, the key conclusion for performing differential
synthesis analysis is that:</p>
<p><span class="math display">\[
L2FC(ksyn) = L2FC(RNA) + L2FC(kdeg)
\]</span></p>
<p>bakR provides <span class="math inline">\(L2FC(kdeg)\)</span> and
DESeq2 provides <span class="math inline">\(L2FC(RNA)\)</span>, so we
have everything we need to calculate the <span class="math inline">\(L2FC(ksyn)\)</span>. Let’s make a data.frame with
this information:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ksyn_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>L2FC <span class="op">=</span> <span class="va">reso</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">+</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">L2FC_kdeg</span>,</span>
<span>                      Gene <span class="op">=</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">XF</span><span class="op">)</span></span>
<span>                </span></code></pre></div>
<p>We are missing one key component though: the uncertainty (or standard
error). How can we get that from bakR and DESeq2? Well the easiest and
most conservative thing to do is to assume independence of <span class="math inline">\(L2FC(RNA)\)</span> and <span class="math inline">\(L2FC(kdeg)\)</span>. That’s because the variances
of independent random variables add, so that:</p>
<p><span class="math display">\[
\begin{align}
\text{Var}[L2FC(ksyn)] &amp;= \text{Var}[L2FC(kdeg) + L2FC(RNA)] \\
&amp;= \text{Var}[L2FC(kdeg)] + \text{Var}[L2FC(RNA)]
\end{align}
\]</span></p>
<p>where <span class="math inline">\(\text{Var}[X]\)</span> is the
variance of random variable <span class="math inline">\(X\)</span>. We
can then calculate the standard error as the square root of the
variance, meaning:</p>
<p><span class="math display">\[
\text{se}[L2FC(ksyn)] = \sqrt{se[L2FC(kdeg)]^2 + se[L2FC(RNA)]^2}
\]</span> where I am using the non-standard notation <span class="math inline">\(\text{se}[X]\)</span> to mean the standard error
of <span class="math inline">\(X\)</span>. With this, we can fill out
the final component of the L2FC(ksyn) dataframe:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ksyn_df</span><span class="op">$</span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">reso</span><span class="op">$</span><span class="va">lfcSE</span><span class="op">^</span><span class="fl">2</span> <span class="op">+</span> <span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">se</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">)</span> </span>
<span>                </span></code></pre></div>
<p>The se in Effects_df is for the log base-e fold change in <span class="math inline">\(kdeg\)</span>, so we have to use the fact
that:</p>
<p><span class="math display">\[
\text{Var}[a*X] = a^2*\text{Var}[X]
\]</span> for some constant a. In this case, a is <span class="math inline">\(log_2(e)\)</span> because that is the
multiplicative factor necessary to convert from log base-2 to log
base-e. Now all that is left is to calculate the p-value and
multiple-test adjust it (I am going to use Benjamini-Hochberg for
multiple test adjustment because that is what both bakR and DESeq2
use).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate p-value using asymptotic Wald test</span></span>
<span><span class="va">ksyn_df</span> <span class="op">&lt;-</span> <span class="va">ksyn_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pval <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">L2FC</span><span class="op">/</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         padj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span><span class="op">)</span></span>
<span>                </span></code></pre></div>
<p>You can now make a <span class="math inline">\(L2FC(ksyn)\)</span>
volcano plot to visualize the final conclusions!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Add conclusion at 0.01 FDR control</span></span>
<span><span class="va">ksyn_df</span> <span class="op">&lt;-</span> <span class="va">ksyn_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>conclusion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.01</span>, </span>
<span>                             <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">L2FC</span> <span class="op">&lt;</span> <span class="fl">0</span>, <span class="st">"Decreased txn"</span>, <span class="st">"Increased txn"</span><span class="op">)</span>,</span>
<span>                             <span class="st">"Not sig."</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make volcano plot</span></span>
<span><span class="va">ksyn_volc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ksyn_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">L2FC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">conclusion</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"L2FC(ksyn)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10(padj)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"orange"</span>, <span class="st">"gray"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Observe volcano plot</span></span>
<span><span class="va">ksyn_volc</span></span></code></pre></div>
<p><img src="Further-Analyses_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The simulation is such that features number 1-800 will be the
non-differentially synthesized RNA and features 801-1000 will be the 200
differentially synthesized RNA. Therefore, we can also see how the
analysis performed</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Add conclusion at 0.01 FDR control</span></span>
<span><span class="va">ksyn_df</span> <span class="op">&lt;-</span> <span class="va">ksyn_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>result <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.01</span>, </span>
<span>                             <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">800</span>, <span class="st">"FP"</span>, <span class="st">"TP"</span><span class="op">)</span>,</span>
<span>                             <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">800</span>, <span class="st">"TN"</span>, <span class="st">"FN"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make volcano plot</span></span>
<span><span class="va">ksyn_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ksyn_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">L2FC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">result</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"L2FC(ksyn)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10(padj)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"gray"</span>, <span class="st">"forestgreen"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Observe volcano plot</span></span>
<span><span class="va">ksyn_results</span></span></code></pre></div>
<p><img src="Further-Analyses_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<p>FP is a false positive, FN is a false negative, TP is a true
positive, and TN is a true negative. All in all, I’d say it went pretty
well! You now know how to perform differential synthesis analysis with
bakR + DEseq2.</p>
</div>
<div class="section level2">
<h2 id="going-beyond-the-steady-state-assumption">Going beyond the steady-state assumption<a class="anchor" aria-label="anchor" href="#going-beyond-the-steady-state-assumption"></a>
</h2>
<p>You may have noticed an assumption that continually crops up in the
analyses presented in bakR vignettes: steady-state. In the
Getting-Started vignette, it was a necessary part of relating the
fraction new to the more interesting <span class="math inline">\(kdeg\)</span>. In this vignette, it was
additionally a key part of assessing synthesis kinetics. It therefore
underlies all of the differential kinetic analysis discussed so far.
What happens if it is a bad assumption? “Dogs and cats living together!
Mass hysteria!”? No Dr. Venkman, there’s a way out, and that’s what
we’re going to talk about next.</p>
<div class="section level3">
<h3 id="violating-the-steady-state-assumption">Violating the steady-state assumption<a class="anchor" aria-label="anchor" href="#violating-the-steady-state-assumption"></a>
</h3>
<p>Before we talk about what to do when the steady-state assumption is
questionable, it’s worth briefly mentioning <em>when</em> the
steady-state assumption is worth throwing out. Assuming steady-state RNA
dynamics means that you are assuming the rate at which RNA synthesized
is the same as the rate at which it is degraded. “Wait, does that mean
we are assuming <span class="math inline">\(ksyn = kdeg\)</span>?”
No. The degradation rate constant <span class="math inline">\(kdeg\)</span> is not the rate of degradation. The
rate of degradation (i.e., the number of RNA molecules degraded per unit
time) depends on <span class="math inline">\(kdeg\)</span> AND the
amount of RNA present. So steady-state actually means that <span class="math inline">\(ksyn = kdeg*[RNA]\)</span>. In the context of
NR-seq experiments, assuming steady-state means that we are assuming
this relationship holds throughout the entire metabolic label feed time.
Therefore, another way to phrase this assumption is to say that we are
assuming that our population of cells aren’t actively responding to some
stimulus and regulating <span class="math inline">\(ksyn\)</span> and/or
<span class="math inline">\(kdeg\)</span>. RNA levels remain constant
during the experiment.</p>
<p>Here’s a fun question: are cells ever at steady-state? Yes and no. An
individual cell is NEVER at steady-state. Gene expression is constantly
regulated as the cell progresses through the cell cycle and performs all
of the biochemical functions necessary to stay alive. BUT, bakR is
designed to analyze <em>bulk</em> nucleotide recoding RNA-seq data. This
means the data comes from a population of cells, not a single cell.
Thus, when I say that I am assuming steady-state, what I really mean is
that I am assuming the population average is at steady-state. If the
population of cells are asynchronous (so they aren’t all going through
the exact same cell cycle stages at the same time) and haven’t been
perturbed recently (e.g., treated with a drug), then this is probably a
pretty solid assumption.</p>
<p>Lots of cool experiments are done with cells far from steady-state
though. When using nucleotide recoding RNA-seq, it is increasingly
popular to perform rapid degradation of a protein of interest, treatment
with a specific drug, or some other acute perturbation leading up to and
during metabolic labeling. Therefore, we need a way to analyze data
without the helpful steady-state assumption. This is challenging though,
because if we can’t assume steady-state, how do we relate what we
measure (the fraction new) to the RNA metabolic kinetics we care about?
The answer: keep it non-parametric.</p>
</div>
<div class="section level3">
<h3 id="analyzing-nr-seq-data-far-from-steady-state">Analyzing NR-seq data far from steady-state<a class="anchor" aria-label="anchor" href="#analyzing-nr-seq-data-far-from-steady-state"></a>
</h3>
<p>I am using the term non-parametric here to suggest getting away from
using explicit mathematical functions to model RNA metabolism. Rather,
we can think very generally about what effects changes in RNA synthesis
and degradation kinetics can have on the fraction new in an NR-seq
experiment. Let’s say we compare an unperturbed, steady-state population
of cells to a cells that have been treated with a drug of interest. If
the fraction new for a particular transcript is higher in the drug
treated cells than the untreated cells, how could that have come about?
Well, generally speaking, there are two possibilities:</p>
<ol style="list-style-type: decimal">
<li>The RNA may have been destabilized by the drug treatment. This is
the same explanation you would give for the change if everything were
steady-state.</li>
<li>The synthesis rate of the RNA may have been ramped up following drug
treatment. This is an intuitive explanation that is surprisingly only
possible away from steady-state. At steady-state, changes in synthesis
kinetics affect the amount of new and old RNA equally, leading to no
change in the fraction new.</li>
</ol>
<p>Here is the key realization: these two possibilities have different
impacts on the LEVELS of RNA. If synthesis is increasing upon drug
treatment, then the amount of that transcript will also increase. If the
transcript is destabilized though, then there will be less of it around
at the end of the experiment. Therefore, <strong>combining differential
expression with differential fraction new analyses reveals the kinetic
differences</strong>. More generally, here is a diagram that covers the
range of possible kinetic mechanisms and conclusions:</p>
<div class="figure">
<img src="NSS_Analysis.png" alt=""><p class="caption">Diagram of NR-seq NSS analysis strategy.</p>
</div>
</div>
<div class="section level3">
<h3 id="performing-steady-state-independent-comparisons-with-bakr-deseq2">Performing steady-state independent comparisons with bakR +
DESeq2<a class="anchor" aria-label="anchor" href="#performing-steady-state-independent-comparisons-with-bakr-deseq2"></a>
</h3>
<p>The goal of the steady-state independent analysis that I am putting
forth is to identify the <strong>major</strong> kinetic mechanism of any
differential expression observed. The analysis strategy is as
follows:</p>
<ol style="list-style-type: decimal">
<li>Compare NR-seq with bakR (but with one small parameter tweak)</li>
<li>Perform differential expression analysis</li>
<li>Use the NSSHeat() function in bakR to combine the results and come
to a mechanistic conclusion.</li>
</ol>
<p>Simple enough! As usual, let’s start out by simulating some data.
Unfortunately, since non-steady state (NSS) is such a broad term, there
is no NSS simulation option. That’s okay though, as the analysis
strategy will still work even if the steady-state assumption isn’t
invalid. To make it interesting though, we’ll simulate a mix of
synthesis and degradation changes:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_bakRData.html">Simulate_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>,</span>
<span>                         num_kd_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>                         num_ks_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 1000 features, 2 experimental conditions</span></span>
<span>  <span class="co"># and 3 replicates for each experimental condition.</span></span>
<span>  <span class="co"># See ?Simulate_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Extract simulated bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span></span>
<span></span>
<span><span class="co"># Extract simualted ground truths</span></span>
<span><span class="va">sim_truth</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">sim_list</span></span>
<span></span>
<span><span class="co">## Run the efficient model</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bakRData</span>, NSS <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating labeled mutation rate</span></span>
<span><span class="co">#&gt; Estimated pnews for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.050<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     1     3 0.050<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     1 0.050<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     2     2 0.050<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     2     3 0.049<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate</span></span>
<span><span class="co">#&gt; Estimated pold is:  0.001</span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>What’s interesting about this simulation is that all transcripts
which are differentially stable are also differentially synthesized, so
our goal is to identify the major source of kinetic regulation. Also,
note the change to the bakRFit() call; a parameter called NSS is set to
TRUE. This will make it so that bakR compares logit(fraction new)s
rather than log(kdeg)s, since the relationship between fraction new and
<span class="math inline">\(kdeg\)</span> may no longer hold.</p>
<p>Next, we can perform differential expression analysis just as we did
before:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Get the count matrix from bakR</span></span>
<span><span class="va">Counts</span> <span class="op">&lt;-</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Data_lists</span><span class="op">$</span><span class="va">Count_Matrix</span></span>
<span></span>
<span><span class="co"># Experimental conditions for each sample</span></span>
<span><span class="co"># There are 6 s4U treated samples (3 replicates of each condition)</span></span>
<span><span class="co"># In addition, there are 2 -s4U control samples (1 for each condition)</span></span>
<span></span>
<span><span class="co">## s4U conditions</span></span>
<span><span class="co"># 1st three samples are reference (ref) samples</span></span>
<span><span class="co"># Next three samples are experimental (exp) samples</span></span>
<span><span class="va">conditions_s4U</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## -s4U control conditions</span></span>
<span><span class="co"># 1st sample is reference, next is experimental</span></span>
<span><span class="va">conditions_ctl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined s4U and -s4U control conditions</span></span>
<span><span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">conditions_s4U</span>, <span class="va">conditions_ctl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make the colData input for DESeq2</span></span>
<span><span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>conditions <span class="op">=</span> <span class="va">conditions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">colData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make DESeq2 data object</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">Counts</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">colData</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span><span class="va">conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span>
<span></span>
<span><span class="co"># Fit DESeq2 model</span></span>
<span><span class="va">ddso</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Extract results of experimental vs. reference comparison</span></span>
<span><span class="va">reso</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">ddso</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"conditions"</span>, <span class="st">"exp"</span>, <span class="st">"ref"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We’re almost ready to use bakR’s special NSSHeat() function.
NSSHeat() has two required inputs:</p>
<ol style="list-style-type: decimal">
<li>A bakR fit object (preferably fit with NSS set to TRUE)</li>
<li>A data frame containing differential expression analysis
results</li>
</ol>
<p>In particular, the differential expression data frame must contain
four columns named as follows:</p>
<ol style="list-style-type: decimal">
<li>
<strong>XF</strong>: Name of the feature (gene, exon, etc.) that the
read comes from.</li>
<li>
<strong>log2FoldChange</strong>: L2FC(RNA) estimated by the
differential expression analysis.</li>
<li>
<strong>stat</strong>: L2FC(RNA)/se[L2FC(RNA)]; so like a
differential expression z-score.</li>
<li>
<strong>padj</strong>: Multiple test adjusted p-value from
differential expression analysis.</li>
</ol>
<p>Conveniently, our DESeq2 results object has all the information we
need! We can make the differential expression data frame like so:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert to data frame</span></span>
<span><span class="va">reso</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">reso</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make data frame</span></span>
<span><span class="va">DE_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>XF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">reso</span><span class="op">)</span>,</span>
<span>                    log2FoldChange <span class="op">=</span> <span class="va">reso</span><span class="op">$</span><span class="va">log2FoldChange</span>,</span>
<span>                    stat <span class="op">=</span> <span class="va">reso</span><span class="op">$</span><span class="va">stat</span>,</span>
<span>                    padj <span class="op">=</span> <span class="va">reso</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span></span></code></pre></div>
<p>We are ready to perform the steady-state independent analysis
now!</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">NSSheatmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NSSHeat.html">NSSHeat</a></span><span class="op">(</span><span class="va">Fit</span>, <span class="va">DE_df</span><span class="op">)</span></span></code></pre></div>
<p>See ?NSSHeat for information about the additional parameters that can
be specified. Most importantly, there is the bakRModel parameter, which
tells NSSHeat which fit to use in the bakRFit model, if multiple exist.
The default setting is to just use the MLE fit, which should always be
there.</p>
<p>So what can you do with the output of NSSHeat? Well, as the name
implies, this function is specifically designed to help you make
heatmaps. NSSheatmap will have three columns:</p>
<ol style="list-style-type: decimal">
<li>
<strong>bakR_score</strong>: The logit(fraction new) change z-score
from bakR</li>
<li>
<strong>DE_score</strong>: The L2FC(RNA) z-score from DESeq2</li>
<li>
<strong>Mech_score</strong>: A mechanism z-score that quantifies the
extent to which significant changes in gene expression are synthesis or
degradation driven. Synthesis driven = positive numbers; degradation
driven = negative numbers.</li>
</ol>
<p>You can quickly make the heatmap using the lovely pheatmap
package:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Nice red to blue color gradient</span></span>
<span>  <span class="co"># Feel free to use any coloring your heart desires</span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#053061"</span>, <span class="st">"#2166AC"</span>, <span class="st">"#4393C3"</span>, <span class="st">"#92C5DE"</span>, </span>
<span>         <span class="st">"#D1E5F0"</span>, <span class="st">"#F7F7F7"</span>, <span class="st">"#FDDBC7"</span>, <span class="st">"#F4A582"</span>, </span>
<span>         <span class="st">"#D6604D"</span>, <span class="st">"#B2182B"</span>, <span class="st">"#67001F"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># NSS heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">NSSheatmap</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="va">col</span><span class="op">)</span></span></code></pre></div>
<p><img src="Further-Analyses_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Each row in the data frame output by NSSHeat corresponds to a feature
that was differentially expressed in the experimental condition.
Features that were not identified as differentially expressed can either
be the result of an underpowered analysis (i.e., they really are
differentially expressed in the two conditions but more
replicates/sequencing depth would have been required to identify this
confidently) or of complicated concomitant modulation of synthesis and
degradation kinetics. This makes non-differentially expressed features
more difficult to analyze with this strategy, which is why bakR focuses
on the differentially expressed features.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
