<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Steady-state quasi-independent mechanistic investigations • bakR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Steady-state quasi-independent mechanistic investigations">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a></li>
    <li><a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a></li>
    <li><a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a></li>
    <li><a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a></li>
    <li><a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a></li>
    <li><a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a></li>
    <li><a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Steady-state quasi-independent mechanistic investigations</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/master/vignettes/NSS.Rmd" class="external-link"><code>vignettes/NSS.Rmd</code></a></small>
      <div class="d-none name"><code>NSS.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>An alternative title considered for the <a href="https://rnajournal.cshlp.org/content/29/7/958.abstract" class="external-link">bakR
      manuscript</a> was “Dissecting the mechanisms of gene expression
      regulation with bakR”. While the actual title is a very
      accurate/literal description of the task bakR accomplishes
      (performing differential stability/synthesis analyses), the
      rejected title better captures the motivation of developing bakR.
      Differential expression analysis with regular RNA-seq + tools like
      DESeq2, edgeR, limma, etc. allows for the identification of
      regulation. NR-seq (TimeLapse-seq, SLAM-seq, TUC-seq, etc.) + bakR
      opens the door the determining whether observed regulation was
      transcriptionally or post-transcriptionally driven.</p>
      <p>In this vignette, I will discuss functionality introduced in
      version 1.0.0 of bakR that explicitly facilitates the kind of
      mechanistic dissection that inspired bakR’s development.
      Implemented in the <code>DissectMechanism</code> function, bakR
      takes a statistical approach to assessing how likely any RNA
      feature is to be undergoing transcriptional or
      post-transcriptional regulation. I will start with a practical
      discussion of how to use <code>DissectMechanism</code> and
      interpret its output. I will commense with a deeper dive into the
      details of what <code>DissectMechanism</code> does.</p>
    </div>
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>These are the packages you are going to have to install and load in
order to do everything presented in this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Packages that are NOT automatically installed when bakR is installed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:bakR':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plotMA</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Packages which are installed when bakR is installed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     count</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'magrittr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     subtract</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set the seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="key-takeaways-from-this-vignette">Key takeaways from this vignette<a class="anchor" aria-label="anchor" href="#key-takeaways-from-this-vignette"></a>
</h2>
<p>The discussion of <code>DissectMechanism</code> presented here is a
bit involved. Thus, I would like to start with a brief summary of what I
hope you understand about the use of <code>DissectMechanism</code> and
the caveats of its analysis strategy:</p>
<ol style="list-style-type: decimal">
<li>
<code>DissectMechanism</code>’s analysis strategy is
“quasi-steady-state-independent”. More specifically, it assumes that at
least one of the population of cells you are comparing is at
steady-state. The other population can be far away from steady-state,
but things get much more complicated if both populations being compared
are dynamically regulating their RNA levels during the experiment.</li>
<li>You have to be weary of the risk of false positives when using
<code>DissectMechanism</code>. The <code>bakR_cutoff</code> parameter in
<code>DissectMechanism</code> determines which type of false positive
(false transcriptional regulation or false post-transcriptional
regulation) is more likely.
<ul>
<li>
<code>bakR_cutoff</code> closer to 1 (it’s maximum value; setting it
to a higher number is identical to setting it to 1) will yield more
post-transcriptional false positives.</li>
<li>Lower <code>bakR_cutoff</code> values will yield more
transcriptional false positives.</li>
<li>Why this is the case is discussed at the end of this vignette in
great detail.</li>
</ul>
</li>
<li>Choose <code>bakR_cutoff</code> based on which type of false
positive will impact you less.
<ul>
<li>Are you looking for high confidence instances of
post-transcriptional regulation? A <code>bakR_cutoff</code> of 0.05 or
lower is a good choice for you to avoid post-transcriptional false
positives. <em>Are you looking for high confidence instances of
transcriptional regulation? Increasing <code>bakR_cutoff</code> to 0.5
or higher is probably a good idea. </em>Are you equally happy to flag
either type of regulation? <code>bakR_cutoff</code>’s default value of
0.3 should work well.</li>
</ul>
</li>
<li>
<code>DissectMechanism</code> was not part of the original bakR
publication. I developed the analysis strategy in collaboration with a
previous undergraduate in the Simon lab (Matthew Saenz) a while back,
but only recently implemented a modified version of our original idea in
bakR. Thus, if you have any questions after reading this vignette, the
best course of action is to post it to the Issues page of the bakR
Github or email me directly (at <a href="mailto:isaac.vock@yale.edu" class="email">isaac.vock@yale.edu</a>) since there is no other public
material on this topic.</li>
<li>If you are analyzing non-steady-state data,
<code>DissectMechanism</code> will work best if the label times (length
of time cells were metabolically labeled for) are the same in the
datasets being compared. <code>DissectMechanism</code> does what it can
to normalize out label time differences, but this strategy is only 100%
rigorous in the steady-state case.</li>
</ol>
</div>
<div class="section level2">
<h2 id="mechanistic-dissections-of-gene-expression-regulation-with-bakr-deseq2">Mechanistic dissections of gene expression regulation with bakR +
DESeq2<a class="anchor" aria-label="anchor" href="#mechanistic-dissections-of-gene-expression-regulation-with-bakr-deseq2"></a>
</h2>
<p>The goal of the <code>DissectMechanism</code> function discussed in
this vignette is to identify the <strong>major</strong> kinetic
mechanism of any differential expression observed. The analysis strategy
is as follows:</p>
<ol style="list-style-type: decimal">
<li>Compare NR-seq with bakR</li>
<li>Perform differential expression analysis</li>
<li>Use the <code><a href="../reference/DissectMechanism.html">DissectMechanism()</a></code> function in bakR to combine
the results and come to a mechanistic conclusion.</li>
</ol>
<p>Simple enough! As usual, let’s start out by simulating some data. I
will simulate two separate datasets, one with exclusively
transcriptional regulation and one with exclusively post-transcriptional
regulation. This will allow us to see how <code>DissectMechanism</code>
performs in these two different regimes:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Only differential synthesis</span></span>
<span></span>
<span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_relative_bakRData.html">Simulate_relative_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1000000</span>,</span>
<span>                         num_ks_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">300</span><span class="op">)</span>,</span>
<span>                         num_kd_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 1000 features, 1 million reads, 2 experimental conditions</span></span>
<span>  <span class="co"># and 3 replicates for each experimental condition. 300 features will be differentially</span></span>
<span>  <span class="co"># transcribed, and there will be no differential stability</span></span>
<span>  <span class="co"># See ?Simulate_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Extract simulated bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span></span>
<span></span>
<span><span class="co">## Run the efficient model</span></span>
<span><span class="va">Fit_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew    pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.049<span style="text-decoration: underline;">9</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     1     3 0.049<span style="text-decoration: underline;">9</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     1 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     2     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     2     3 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span>
<span></span>
<span></span>
<span><span class="co">### Only differential stability</span></span>
<span></span>
<span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_relative_bakRData.html">Simulate_relative_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1000000</span>,</span>
<span>                         num_ks_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                         num_kd_DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">300</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 1000 features, 1 million reads, 2 experimental conditions</span></span>
<span>  <span class="co"># and 3 replicates for each experimental condition. 300 features will be differentially</span></span>
<span>  <span class="co"># transcribed, and there will be no differential stability</span></span>
<span>  <span class="co"># See ?Simulate_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Extract simulated bakRData object</span></span>
<span><span class="va">bakRData</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span></span>
<span></span>
<span><span class="co">## Run the efficient model</span></span>
<span><span class="va">Fit_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew     pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.050<span style="text-decoration: underline;">0</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">0</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     1     3 0.049<span style="text-decoration: underline;">9</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     1 0.050<span style="text-decoration: underline;">0</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     2     2 0.050<span style="text-decoration: underline;">2</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     2     3 0.050<span style="text-decoration: underline;">1</span> 0.000<span style="text-decoration: underline;">999</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>Next, we can perform differential expression analysis just as
described in the differential synthesis analysis vignette:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Only differential synthesis dataset</span></span>
<span></span>
<span><span class="co"># Get the count matrix from bakR</span></span>
<span><span class="va">Counts</span> <span class="op">&lt;-</span> <span class="va">Fit_s</span><span class="op">$</span><span class="va">Data_lists</span><span class="op">$</span><span class="va">Count_Matrix</span></span>
<span></span>
<span><span class="co"># Experimental conditions for each sample</span></span>
<span><span class="co"># There are 6 s4U treated samples (3 replicates of each condition)</span></span>
<span><span class="co"># In addition, there are 2 -s4U control samples (1 for each condition)</span></span>
<span></span>
<span><span class="co">## s4U conditions</span></span>
<span><span class="co"># 1st three samples are reference (ref) samples</span></span>
<span><span class="co"># Next three samples are experimental (exp) samples</span></span>
<span><span class="va">conditions_s4U</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## -s4U control conditions</span></span>
<span><span class="co"># 1st sample is reference, next is experimental</span></span>
<span><span class="va">conditions_ctl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="st">"exp"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined s4U and -s4U control conditions</span></span>
<span><span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">conditions_s4U</span>, <span class="va">conditions_ctl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make the colData input for DESeq2</span></span>
<span><span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>conditions <span class="op">=</span> <span class="va">conditions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">colData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make DESeq2 data object</span></span>
<span><span class="va">dds_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">Counts</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">colData</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span><span class="va">conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span>
<span></span>
<span><span class="co"># Fit DESeq2 model</span></span>
<span><span class="va">ddso_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds_s</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Extract results of experimental vs. reference comparison</span></span>
<span><span class="va">reso_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">ddso_s</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"conditions"</span>, <span class="st">"exp"</span>, <span class="st">"ref"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### Only differential stability dataset</span></span>
<span></span>
<span><span class="co"># Get the other count matrix from bakR</span></span>
<span><span class="va">Counts</span> <span class="op">&lt;-</span> <span class="va">Fit_d</span><span class="op">$</span><span class="va">Data_lists</span><span class="op">$</span><span class="va">Count_Matrix</span></span>
<span></span>
<span><span class="co"># Make DESeq2 data object</span></span>
<span><span class="va">dds_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="va">Counts</span>,</span>
<span>                              colData <span class="op">=</span> <span class="va">colData</span>,</span>
<span>                              design <span class="op">=</span> <span class="op">~</span><span class="va">conditions</span><span class="op">)</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span>
<span></span>
<span><span class="co"># Fit DESeq2 model</span></span>
<span><span class="va">ddso_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds_d</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Extract results of experimental vs. reference comparison</span></span>
<span><span class="va">reso_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">ddso_d</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"conditions"</span>, <span class="st">"exp"</span>, <span class="st">"ref"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We’re almost ready to use bakR’s special DissectMechanism() function.
DissectMechanism() has two required inputs:</p>
<ol style="list-style-type: decimal">
<li>A bakR fit object</li>
<li>A data frame containing differential expression analysis
results</li>
</ol>
<p>In particular, the differential expression data frame must contain
five columns named as follows:</p>
<ol style="list-style-type: decimal">
<li>
<strong>XF</strong>: Name of the feature (gene, exon, etc.) that the
read comes from.</li>
<li>
<strong>L2FC_RNA</strong>: L2FC(RNA) estimated by the differential
expression analysis.</li>
<li>
<strong>DE_score</strong>: Differential expression z-score.</li>
<li>
<strong>DE_se</strong>: L2FC(RNA) standard error.</li>
<li>
<strong>DE_pval</strong>: P-value from differential expression
analysis.</li>
<li>
<strong>DE_padj</strong>: Multiple test adjusted p-value from
differential expression analysis.</li>
</ol>
<p>Conveniently, our DESeq2 results object has all the information we
need! We can make the differential expression data frame like so:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Only differential synthesis: </span></span>
<span></span>
<span><span class="co"># Convert to data frame</span></span>
<span><span class="va">reso_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">reso_s</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make data frame</span></span>
<span><span class="va">DE_df_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>XF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">reso_s</span><span class="op">)</span>,</span>
<span>                    L2FC_RNA <span class="op">=</span> <span class="va">reso_s</span><span class="op">$</span><span class="va">log2FoldChange</span>,</span>
<span>                    DE_score <span class="op">=</span> <span class="va">reso_s</span><span class="op">$</span><span class="va">stat</span>,</span>
<span>                    DE_se <span class="op">=</span> <span class="va">reso_s</span><span class="op">$</span><span class="va">lfcSE</span>,</span>
<span>                    DE_pval <span class="op">=</span> <span class="va">reso_s</span><span class="op">$</span><span class="va">pval</span>,</span>
<span>                    DE_padj <span class="op">=</span> <span class="va">reso_s</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### Only differential degradation:</span></span>
<span></span>
<span><span class="co"># Convert to data frame</span></span>
<span><span class="va">reso_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">reso_d</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make data frame</span></span>
<span><span class="va">DE_df_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>XF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">reso_d</span><span class="op">)</span>,</span>
<span>                    L2FC_RNA <span class="op">=</span> <span class="va">reso_d</span><span class="op">$</span><span class="va">log2FoldChange</span>,</span>
<span>                    DE_score <span class="op">=</span> <span class="va">reso_d</span><span class="op">$</span><span class="va">stat</span>,</span>
<span>                    DE_se <span class="op">=</span> <span class="va">reso_d</span><span class="op">$</span><span class="va">lfcSE</span>,</span>
<span>                    DE_pval <span class="op">=</span> <span class="va">reso_d</span><span class="op">$</span><span class="va">pval</span>,</span>
<span>                    DE_padj <span class="op">=</span> <span class="va">reso_d</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span></span></code></pre></div>
<p>We are ready to perform a mechanistic dissection now!</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Decreasing sims parameter to speed up; wouldn't normally suggest this if you</span></span>
<span><span class="co"># want higher precision mechanism p-values, discussed later</span></span>
<span><span class="va">Mechs_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DissectMechanism.html">DissectMechanism</a></span><span class="op">(</span><span class="va">Fit_s</span>, <span class="va">DE_df_s</span>,</span>
<span>                          sims <span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Combining bakR and DE analyses</span></span>
<span><span class="co">#&gt; Calculating mechanism p-value. This could take several minutes...</span></span>
<span><span class="co">#&gt; Assessing differential synthesis</span></span>
<span><span class="co">#&gt; Constructing Heatmap_df</span></span>
<span></span>
<span><span class="va">Mechs_d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DissectMechanism.html">DissectMechanism</a></span><span class="op">(</span><span class="va">Fit_d</span>, <span class="va">DE_df_d</span>,</span>
<span>                          sims <span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Combining bakR and DE analyses</span></span>
<span><span class="co">#&gt; Calculating mechanism p-value. This could take several minutes...</span></span>
<span><span class="co">#&gt; Assessing differential synthesis</span></span>
<span><span class="co">#&gt; Constructing Heatmap_df</span></span></code></pre></div>
<p>See <code><a href="../reference/DissectMechanism.html">?DissectMechanism</a></code> for information about the
additional parameters that can be specified. Most importantly, there is
the bakRModel parameter, which tells DissectMechanism which fit to use
in the bakRFit model, if multiple exist. The default setting is to just
use the MLE fit, which should always be there.</p>
<p>The output of <code>DissectMechanism</code> is two data frames:</p>
<ol style="list-style-type: decimal">
<li>Heatmap_df: A data frame meant to facilitate visualization of
results as a heatmap.</li>
<li>Mechanism_df: A data frame with statistical assessments of likely
mechanisms of observed differential expression.</li>
</ol>
<p>Heatmap_df has the three following columns:</p>
<ol style="list-style-type: decimal">
<li>
<strong>bakR_score</strong>: The logit(fraction new) change z-score
from bakR</li>
<li>
<strong>DE_score</strong>: The L2FC(RNA) z-score from DESeq2</li>
<li>
<strong>Mech_score</strong>: A mechanism z-score that quantifies the
extent to which significant changes in gene expression are synthesis or
degradation driven. Synthesis driven = positive numbers; degradation
driven = negative numbers.</li>
</ol>
<p>You can quickly make the heatmap using the lovely pheatmap
package:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Nice red to blue color gradient</span></span>
<span>  <span class="co"># Feel free to use any coloring your heart desires</span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#053061"</span>, <span class="st">"#2166AC"</span>, <span class="st">"#4393C3"</span>, <span class="st">"#92C5DE"</span>, </span>
<span>         <span class="st">"#D1E5F0"</span>, <span class="st">"#F7F7F7"</span>, <span class="st">"#FDDBC7"</span>, <span class="st">"#F4A582"</span>, </span>
<span>         <span class="st">"#D6604D"</span>, <span class="st">"#B2182B"</span>, <span class="st">"#67001F"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Heatmap for differential synthesis only dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">Mechs_s</span><span class="op">$</span><span class="va">Heatmap_df</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="va">col</span><span class="op">)</span></span></code></pre></div>
<p><img src="NSS_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Heatmap for differential degradation only dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">Mechs_d</span><span class="op">$</span><span class="va">Heatmap_df</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="va">col</span><span class="op">)</span></span></code></pre></div>
<p><img src="NSS_files/figure-html/unnamed-chunk-6-2.png" width="576" style="display: block; margin: auto;"></p>
<p>Each row in the data frame output by DissectMechanism corresponds to
a feature that was differentially expressed in the experimental
condition. Differential expression is defined as a <code>DE_padj</code>
less than the <code>DE_cutoff</code> parameter in
<code>DissectMechanism</code>. These are the features with the highest
confidence observed regulation, and thus the ones you are most likely to
be interested in assessing the mechanism of regulation.</p>
<p>Mechanism_df provides all of the information present in Heatmap_df,
and then some, for all features, even those that are not “significantly
differentially expressed”. The columns of Mechanism_df are as
follows:</p>
<ol style="list-style-type: decimal">
<li>
<strong>XF</strong>: Feature name</li>
<li>
<strong>bakR_score</strong>: bakR z-score for kdeg difference</li>
<li>
<strong>L2FC_kdeg</strong>: L2FC(kdeg) estimated by bakR</li>
<li>
<strong>bakR_pval</strong>: P-value of L2FC(kdeg) estimated by
bakR</li>
<li>
<strong>bakR_padj</strong>: bakR’s multiple test adjusted p
value</li>
<li>
<strong>L2FC_RNA</strong>: L2FC([RNA]) estimated by a differential
expression analysis (DEA) tool</li>
<li>
<strong>DE_score</strong>: L2FC([RNA]) z-score from DEA tool</li>
<li>
<strong>DE_pval</strong>: P-value for differential expression</li>
<li>
<strong>DE_padj</strong>: Multiple test adjusted p-value for
differential expression</li>
<li>
<strong>mech_stat</strong>: Mechansism assignment test
statistic</li>
<li>
<strong>mech_pval</strong>: P-value for mechanism assignment</li>
<li>
<strong>mech_padj</strong>: Multiple test adjusted p-value for
mechanism assignment</li>
<li>
<strong>meta_pval</strong>: P-value that either bakR or DEA z-score
is non-null (i.e., a bakR + DEA meta analysis p-value)</li>
<li>
<strong>meta_padj</strong>: Multiple test adjusted meta analysis
p-value</li>
<li>
<strong>L2FC_ksyn</strong>: L2FC(ksyn) estimated as discussed in the
differential synthesis analysis vignette.</li>
<li>
<strong>ksyn_score</strong>: L2FC(ksyn) z-score estimated as
discusssed in the differential synthesis analysis vignette.</li>
<li>
<strong>ksyn_pval</strong>: L2FC(ksyn) p-value estimated as
discussed in the differential synthesis analysis vignette.</li>
<li>
<strong>ksyn_padj</strong>: Multiple test adjusted L2FC(ksyn)
p-value.</li>
<li>
<strong>f_deg</strong>: Fraction of L2FC(RNA) attributable to
L2FC(kdeg). Equals -L2FC(kdeg)/L2FC(RNA) (with a ceiling of 1 and a
floor of 0), and is an old steady-state dependent metric our lab used.
Values greater than 0.5 represent degradation driven regulation; values
less than 0.5 represent synthesis driven regulation.</li>
</ol>
<p>The first nine columns are things you provided to
<code>DissectMechanism</code>. The next five columns are the important
output of <code>DissectMechanism</code>’s statistical analysis. This
analysis really consists of two separate analyses:</p>
<ol style="list-style-type: decimal">
<li>Mechanistic analysis: Assessing whether or not there is evidence for
transcriptionally or post-transcriptionally driven gene expression
regulation.
<ul>
<li>As discussed above, positive values for <strong>mech_stat</strong>
represent transcriptional (i.e., synthesis-driven) regulation. Negative
values represent post-transcriptional (i.e., stability-driven)
regulation.</li>
</ul>
</li>
<li>Meta-analysis: Assessing whether or not there is evidence for any
of: differential expression, differential kinetics (bakR output), or
both.
<ul>
<li>For this, Fisher’s method is used to combine the DEA and bakR
p-values and assess the probability of obtaining a duo of p-values like
this if both the differential expression and differential fraction new
signal were null.</li>
</ul>
</li>
</ol>
<p>The next 4 columns are the output of differential synthesis analysis
conducted as described in the vignette on this topic. For details, check
out that vignette. Finally, the last column is a metric that our lab has
used in the past to assess the mechanisms of observed regulation. It is
only valid if your population of cells are at steady-state, as it relies
on the relationship that arises from this assumption: L2FC(RNA) =
L2FC(ksyn) - L2FC(kdeg).</p>
<p>One thing I like to look at is how the combination of the estimated
L2FC(RNA) and L2FC(kdeg) give rise to the mechanism statistic. Let’s
look at the dataset with exclusively transcriptional regulation:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Scatter plot of L2FC(kdeg) vs. L2FC(RNA) colored by mechanism test stat</span></span>
<span>  <span class="co"># Gotta transform the mech_stat because it spans many orders of magnitude</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">Mechs_s</span><span class="op">$</span><span class="va">Mechanism_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">L2FC_kdeg</span>, y <span class="op">=</span> <span class="va">L2FC_RNA</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">mech_stat</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">mech_stat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"L2FC(kdeg) from bakR"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"L2FC(RNA) from DESeq2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Mechanism"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NSS_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>and next for the dataset with exclusively post-transcriptional
regulation</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Scatter plot of L2FC(kdeg) vs. L2FC(RNA) colored by mechanism test stat</span></span>
<span>  <span class="co"># Gotta transform the mech_stat because it spans many orders of magnitude</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">Mechs_d</span><span class="op">$</span><span class="va">Mechanism_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">L2FC_kdeg</span>, y <span class="op">=</span> <span class="va">L2FC_RNA</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">mech_stat</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">mech_stat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"L2FC(kdeg) from bakR"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"L2FC(RNA) from DESeq2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Mechanism"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NSS_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>If you take a closer look at the heatmap and scatter plot, a couple
things might stick out:</p>
<ol style="list-style-type: decimal">
<li>In the case of all differential expression being driven by
transcriptional regulation, there were no false positives.
<code>DissectMechanism</code> correctly classified all instances of
differential expression as transcriptionally driven.</li>
<li>In the case of all differential expression being driven by RNA
stability regulation, there were a patch of false positives (features
identified as transcriptionally regulated). Still,
<code>DissectMechanism</code> did a decent job, though I will have more
to say about the challenges of this analysis in the next section.</li>
<li>Higher L2FC(RNA) and L2FC(kdeg) yields higher mechanism scores, but
it is possible to have a high, positive mechanism score in the absence
of a large L2FC(kdeg). The scoring function will be discussed in more
detail in the next section, so check that out if you are
interested.</li>
</ol>
</div>
<div class="section level2">
<h2 id="going-beyond-the-steady-state-assumption">Going beyond the steady-state assumption<a class="anchor" aria-label="anchor" href="#going-beyond-the-steady-state-assumption"></a>
</h2>
<p>You may have noticed an assumption that continually crops up in the
analyses presented in bakR vignettes: steady-state. In the
Getting-Started vignette, it was a necessary part of relating the
fraction new to the biologically relevant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">kdeg</annotation></semantics></math>.
In the differential synthesis analysis vignette it was a key part of
assessing synthesis kinetics. It therefore underlies all of the
differential kinetic analysis discussed so far. What happens if it is a
bad assumption? “Dogs and cats living together! Mass hysteria!”? Well
Dr. Venkman, it certainly complicates things, but
<code>DissectMechanism</code> does its best to curtail these issues.</p>
<p>In this section, I will discuss many of the nitty-gritty details
about how <code>DissectMechanism</code> is working. It’s analysis
strategy is heavily inspired by the task of assessing the kinetic
mechanisms of observed differential expression in a
steady-state-independent manner. Accomplishing this means generalizing
how differences in the fraction new in an NR-seq experiment can relate
to differences in degradation <strong>or</strong> synthesis when away
from steady-state.</p>
<div class="section level3">
<h3 id="violating-the-steady-state-assumption">Violating the steady-state assumption<a class="anchor" aria-label="anchor" href="#violating-the-steady-state-assumption"></a>
</h3>
<p>Before we talk about what to do when the steady-state assumption is
questionable, it’s worth briefly mentioning <em>when</em> the
steady-state assumption is worth throwing out. Assuming steady-state RNA
dynamics means that you are assuming the rate at which RNA synthesized
is the same as the rate at which it is degraded. “Wait, does that mean
we are assuming
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>s</mi><mi>y</mi><mi>n</mi><mo>=</mo><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">ksyn = kdeg</annotation></semantics></math>?”
No. The degradation rate constant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">kdeg</annotation></semantics></math>
is not the rate of degradation. The rate of degradation (i.e., the
number of RNA molecules degraded per unit time) depends on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">kdeg</annotation></semantics></math>
AND the amount of RNA present. So steady-state actually means that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>s</mi><mi>y</mi><mi>n</mi><mo>=</mo><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi><mo>*</mo><mrow><mo stretchy="true" form="prefix">[</mo><mi>R</mi><mi>N</mi><mi>A</mi><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">ksyn = kdeg*[RNA]</annotation></semantics></math>.
In the context of NR-seq experiments, assuming steady-state means that
we are assuming this relationship holds throughout the entire metabolic
label feed time. Therefore, another way to phrase this assumption is to
say that we are assuming that our population of cells aren’t actively
responding to some stimulus and regulating
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>s</mi><mi>y</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">ksyn</annotation></semantics></math>
and/or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>d</mi><mi>e</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">kdeg</annotation></semantics></math>.
RNA levels remain constant during the experiment.</p>
<p>Here’s a fun question: are cells ever at steady-state? Yes and no. An
individual cell is almost never at steady-state. Gene expression is
constantly regulated as the cell progresses through the cell cycle and
performs all of the biochemical functions necessary to stay alive. BUT,
bakR is designed to analyze <em>bulk</em> nucleotide recoding RNA-seq
data. This means the data comes from a population of cells, not a single
cell. Thus, when I say that I am assuming steady-state, what I really
mean is that I am assuming the population average is at steady-state. If
the population of cells are asynchronous (so they aren’t all going
through the exact same cell cycle stages at the same time) and haven’t
been perturbed recently (e.g., treated with a drug), then this is
probably a pretty solid assumption.</p>
<p>Lots of cool experiments are done with cells far from steady-state
though. When using nucleotide recoding RNA-seq, it is increasingly
popular to perform rapid degradation of a protein of interest, treatment
with a specific drug, or some other acute perturbation leading up to and
during metabolic labeling. Therefore, we need a way to analyze data
without the helpful steady-state assumption. This is challenging though,
because if we can’t assume steady-state, how do we relate what we
measure (the fraction new) to the RNA metabolic kinetics we care about?
The answer: keep it non-parametric.</p>
</div>
<div class="section level3">
<h3 id="analyzing-nr-seq-data-far-from-steady-state">Analyzing NR-seq data far from steady-state<a class="anchor" aria-label="anchor" href="#analyzing-nr-seq-data-far-from-steady-state"></a>
</h3>
<p>I am using the term non-parametric here to suggest getting away from
using explicit mathematical functions to model RNA metabolism. Rather,
we can think very generally about what effects changes in RNA synthesis
and degradation kinetics can have on the fraction new in an NR-seq
experiment. Let’s say we compare an unperturbed, steady-state population
of cells to a cells that have been treated with a drug of interest. If
the fraction new for a particular transcript is higher in the drug
treated cells than the untreated cells, how could that have come about?
Well, generally speaking, there are two possibilities:</p>
<ol style="list-style-type: decimal">
<li>The RNA may have been destabilized by the drug treatment. This is
the same explanation you would give for the change if everything were
steady-state.</li>
<li>The synthesis rate of the RNA may have been ramped up following drug
treatment. This is an intuitive explanation that is surprisingly only
possible away from steady-state. At steady-state, changes in synthesis
kinetics affect the amount of new and old RNA equally, leading to no
change in the fraction new.</li>
</ol>
<p>Here is the key realization: these two possibilities have different
impacts on the LEVELS of RNA. If synthesis is increasing upon drug
treatment, then the amount of that transcript will also increase. If the
transcript is destabilized though, then there will be less of it around
at the end of the experiment. Therefore, <strong>combining differential
expression with differential fraction new analyses reveals the kinetic
differences</strong>. More generally, here is a diagram that covers the
range of possible kinetic mechanisms and conclusions
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>
is the fraction new):</p>
<div class="float">
<img src="NSS_Analysis.png" alt="Diagram of NR-seq NSS analysis strategy."><div class="figcaption">Diagram of NR-seq NSS analysis strategy.</div>
</div>
</div>
<div class="section level3">
<h3 id="the-statistics-of-dissectmechanism">The statistics of DissectMechanism<a class="anchor" aria-label="anchor" href="#the-statistics-of-dissectmechanism"></a>
</h3>
<p><code>DissectMechanism</code> attempts to formalize the analysis
discussed in the last section. It combines a differential fraction new
test statistic with a differential expression test statistic. The
differential fraction new test statistic is just the L2FC(kdeg) z-score
output by bakR. Isn’t this invalidated by violations of the steady-state
assumption? Yes and no. While it is true that bakR’s estimated
L2FC(kdeg) may no longer truly represent the change in the degradation
rate constant, it still does represent an accurate assessment of the
change in the fraction new. The kdeg is estimated as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>k</mi><mtext mathvariant="normal">deg</mtext></msub><mo>=</mo><mo>−</mo><mtext mathvariant="normal">ln</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>f</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mtext mathvariant="normal">tl</mtext></mrow><annotation encoding="application/x-tex">
k_{\text{deg}} = -\text{ln}(1 - fn)/\text{tl}
</annotation></semantics></math> where tl is the label time. While the
transformation on the right of this expression may no longer represent
the degradation rate constant away from steady-state, it is simply a
monotonic transformation of the fraction new. Thus, differences in
bakR’s estimated “kdeg” can still be interpreted as differences in the
fraction new, assuming the label times in the conditions being compared
are the same. If the label times differ, this also represents a naive
(though somewhat inaccurate in the non-steady-state case) normalization
for these differences.</p>
<p>The mechanism test statistic used by <code>DissectMechanism</code> is
quite simple. It’s just:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">mech_stat</mtext><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtext mathvariant="normal">L2FC(kdeg) z-score</mtext><mo stretchy="true" form="postfix">]</mo></mrow><mo>*</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtext mathvariant="normal">L2FC(RNA) z-score</mtext><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">
\text{mech_stat} = [\text{L2FC(kdeg) z-score}]*[\text{L2FC(RNA) z-score}]
</annotation></semantics></math></p>
<p>Why this is a reasonable test-statistic choice is best demonstrated
via visualization of this function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Grid to plot on</span></span>
<span></span>
<span><span class="co"># Grid size in each axis</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span></span>
<span><span class="co"># DE z-scores</span></span>
<span><span class="va">DE_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="op">-</span><span class="fl">10</span>, to <span class="op">=</span> <span class="fl">10</span>, length.out <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, times <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># bakR z-scores</span></span>
<span><span class="va">bakR_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="op">-</span><span class="fl">10</span>, to <span class="op">=</span> <span class="fl">10</span>, length.out <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, each <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span>
<span></span>
<span><span class="va">grid_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="va">DE_score</span>,</span>
<span>                  bakR <span class="op">=</span> <span class="va">bakR_score</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Mechanism score</span></span>
<span><span class="va">grid_df</span> <span class="op">&lt;-</span> <span class="va">grid_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Mech <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">DE</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>                       <span class="op">(</span><span class="va">bakR</span> <span class="op">+</span> <span class="fl">2</span><span class="op">)</span><span class="op">*</span><span class="va">DE</span>,</span>
<span>                       <span class="op">(</span><span class="va">bakR</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span><span class="op">*</span><span class="va">DE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate p-value</span></span>
<span><span class="va">grid_df</span> <span class="op">&lt;-</span> <span class="va">grid_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Mech_pval <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Fdist.html" class="external-link">df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Mech</span><span class="op">)</span>, df1 <span class="op">=</span> <span class="fl">2</span>, df2 <span class="op">=</span> <span class="fl">2</span>, ncp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize test statistic</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">grid_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bakR</span>, y <span class="op">=</span> <span class="va">DE</span>, z <span class="op">=</span> <span class="fl">0.03</span><span class="op">*</span><span class="va">Mech</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_contour.html" class="external-link">geom_contour_filled</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"bakR score"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"DE score"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="NSS_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In this plot, dark blue represents high confidence
post-transcriptional regulation and bright yellow represents high
confidence transcriptional regulation. There are four somewhat asymetric
quadrants in this plot:</p>
<p><strong>Top-left</strong>: Fraction new went down and expression
levels went up. This has to be stabilization.</p>
<p><strong>Top-right</strong>: Fraction new went up and expression
levels went up. This has to be transcriptional upregulation.</p>
<p><strong>Bottom-left</strong>: Fraction new went down and expression
went down. Transcriptional downregulation must be at play</p>
<p><strong>Bottom-right</strong>: Fraction new went up and expression
went down. The RNA was destabilized.</p>
<p>This particular function is admittedly a bit arbitrary, but it has a
lot of the trends that you would want in a function attempting to
formalize to the strategy discussed in the last section:</p>
<ol style="list-style-type: decimal">
<li>The quadrants described above match the mechanistic conclusions in
the schematic shown in the last section.</li>
<li>Low confidence in differential fraction new or differential
expression can be balanced out by high confidence in the other. Moderate
confidence in both is as good as high confidence in one and lower
confidence in another.</li>
<li>In keeping with the possibility of steady-state dynamics,
differential expression can be transcriptionally driven without a strong
fraction new signal. More on this in the next section when I discuss the
vertical dividing lines that determine just how strong the fraction new
signal has to be to qualify as “real” signal.</li>
</ol>
</div>
<div class="section level3">
<h3 id="the-assymetry-of-the-mechanism-score">The assymetry of the mechanism score<a class="anchor" aria-label="anchor" href="#the-assymetry-of-the-mechanism-score"></a>
</h3>
<p>I briefly alluded in the last section to the asymmetry of the
mechanism score function. The x-axis (0 differential expression)
represents one dividing line, but what about the two separate vertical
dividing lines? The reason for this is that transcriptional regulation
can yield no fraction new signal. Thus, low confidence changes in the
fraction new should be interpreted as no change in the fraction new.
When no change in the fraction new is accompanied by differential
expression, this is most likely transcriptional regulation.</p>
<p>The locations of the vertical lines are set by the
<code>bakR_cutoff</code> parameter in <code>DissectMechanism</code>.
<code>bakR_cutoff</code> represents the bakR_padj cutoff for calling a
change in the fraction new “significant”. <code>DissectMechanism</code>
determines what z-score is necessary to yield that padj (since
multiple-test adjustment makes the relationship between padj and z-score
a bit difficult to predict a priori) and uses that as the vertical
dividing lines.</p>
<p>The reliance on statistical confidence is an important shortcoming of
<code>DissectMechanism</code>’s analysis strategy. A low confidence
L2FC(kdeg) could be the result of no large changes in the fraction new,
but it could also be the result of low statistical power. In other
words, there may be strong post-transcriptional regulation lurking in
the region between these vertical dividing lines, but the coverage or
number of replicates may have been too low to assess this
confidently.</p>
<p>The consequence of all of this is that <code>bakR_cutoff</code> tunes
the kind of false positives that are more likely to crop up:</p>
<ol style="list-style-type: decimal">
<li>If <code>bakR_cutoff</code> is small, then only large fraction new
z-scores are going to be called significant. Any other low confidence
fraction new changes that are accompanied by differential expression
will be called transcriptional regulation. Thus, small
<code>bakR_cutoff</code>’s yield more transcriptional false
positives.</li>
<li>On the other end of the spectrum, a <code>bakR_cutoff</code> close
to 1 will cause any small change in the fraction new to be interpreted
as a real change. Thus, if the direction of fraction new differences is
opposite the direction of differential expression,
<code>DissectMechanism</code> will call this post-transcriptional
regulation, no matter how inconsequential the fraction new shift. Thus,
large <code>bakR_cutoff</code>’s yield more post-transcriptional false
positives.</li>
</ol>
<p>The conclusion: be wary. As discussed in the summary at the top of
this vignette, <code>bakR_cutoff</code> should be chosen in a way to
limit the false positives that would be more likely to lead you
astray.</p>
</div>
<div class="section level3">
<h3 id="the-mechanism-p-value">The mechanism p-value<a class="anchor" aria-label="anchor" href="#the-mechanism-p-value"></a>
</h3>
<p>The last wrinkle in <code>DissectMechanism</code> is the calculation
of the mechanism score’s p-value. Multiple test adjusted p-values are
nice because they give you a clear sense of how confident you can be
that a given set of mechanistic assignments are accurate. Without these,
it can be difficult to interpret the mechanism scores. It turns out
though that nature of the mechanism score null model makes it difficult
to calculate these p-values with great precision.</p>
<p>Work through a bit of mathematical statistics and you’ll find that
the null hypothesis distribution is the esoteric “doubly non-central
scaled F-distribution”. There is the aptly named <a href="https://CRAN.R-project.org/package=sadists" class="external-link">sadist</a> R package
which implements functions for working with little known distributions,
but even it only accomodates the unscaled version of this distribution.
While one R package exists to calculate integrals of the doubly
non-central scaled F-distribution, the precision of its estimates is
about as good as I was able to shakily code up myself. Therefore, I went
forward with a home brew Monte Carlo approach to p-value calculation. In
short, I simulate from the null model a bunch of times (the number of
which is set by the <code>sims</code> parameter in
<code>DissectMechanism</code>), and calculate the p-value as the
fraction of the time the simulated mechanism score was larger in
magnitude than a given mechanism score.</p>
<p>This means that the precision of the p-value is set by the
<code>sims</code> parameter. By default, it is set to 10 million, which
means that the minimum meaningful p-value it can estimate is
1/10,000,000, or 1e-7. In this vignette, I decreased this by an order of
base-10 magnitude to speed up the runtime, as it can take 10 or more
seconds to collect this many draws from the null. You can get more
precision by increasing <code>sims</code>, but it’s going to cost you
heavily on the runtime.</p>
<p><strong>NOTE</strong>: if 0 of the draws from the null are more
extreme than a particular mechanism score, that mechanism score is
assigned a p-value of 0.5/<code>sims</code>. Because of this, relative
ordering of mechanism scores is not perfectly preserved in the relative
ordering of p-values. Thus, if you are attempting to generate ordered
lists of high confidence mechanistic assignments, order by mech_score,
not mech_pval!!</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
