<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bakR">
<title>Troubleshooting analyses of NR-seq data with bakR • bakR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Troubleshooting analyses of NR-seq data with bakR">
<meta property="og:description" content="bakR">
<meta property="og:image" content="https://simonlabcode.github.io/bakR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bakR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting-Started.html">Differential kinetic analysis with bakR</a>
    <a class="dropdown-item" href="../articles/bakR-Quickstart.html">bakR for people in a hurry</a>
    <a class="dropdown-item" href="../articles/Differential-Synth.html">Differential synthesis analysis with bakR and DESeq2</a>
    <a class="dropdown-item" href="../articles/bakR-Fn.html">GRAND-SLAM output/fn estimates as bakR input</a>
    <a class="dropdown-item" href="../articles/Dropout.html">Correcting for dropout</a>
    <a class="dropdown-item" href="../articles/Troubleshooting.html">Troubleshooting analyses of NR-seq data with bakR</a>
    <a class="dropdown-item" href="../articles/NSS.html">Steady-state quasi-independent mechanistic investigations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/simonlabcode/bakR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Troubleshooting analyses of NR-seq data with bakR</h1>
                        <h4 data-toc-skip class="author">Isaac Vock</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/simonlabcode/bakR/blob/HEAD/vignettes/Troubleshooting.Rmd" class="external-link"><code>vignettes/Troubleshooting.Rmd</code></a></small>
      <div class="d-none name"><code>Troubleshooting.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Analyzing NR-seq data can be difficult. This vignette is here
      to help. In it, I will discuss some of the most common challenges
      that one can run into when analyzing your data with bakR, and how
      to overcome them. In particular, I will showcase bakR’s quality
      control check functionality (introduced in version 0.4.0) and
      discuss the problems that it looks out for. In addition, I will
      discuss aspects of NR-seq data that can hinder accurate estimation
      of the mutation rates in new and old sequencing reads. Finally, I
      will show how bakR’s alternative mutation rate estimation strategy
      can ameliorate some of these challenges.</p>
    </div>
    
<div class="section level2">
<h2 id="necessary-setup">Necessary Setup<a class="anchor" aria-label="anchor" href="#necessary-setup"></a>
</h2>
<p>Load bakR! Version 1.0.0 or later is required to run all of the code
in this vignette and reproduce the results presented. ggplot2 and
corrplot (the latter of which is not automatically installed with bakR)
will also be used for some figures shown here.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://simonlabcode.github.io/bakR/">bakR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; corrplot 0.92 loaded</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-qc_checks-function">The QC_checks() function<a class="anchor" aria-label="anchor" href="#the-qc_checks-function"></a>
</h2>
<p>Version 0.4.0 of bakR introduced a function that takes as input a
bakRFit object and assesses a number of aspects of the model fit and raw
data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_relative_bakRData.html">Simulate_relative_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>, depth <span class="op">=</span> <span class="fl">1000000</span>,</span>
<span>                                       nreps <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 1000 features, 1000000 reads, 2 experimental conditions,</span></span>
<span>  <span class="co"># and 2 replicates for each experimental condition</span></span>
<span>  <span class="co"># See ?Simulate_relative_bakRData for details regarding tunable parameters</span></span>
<span></span>
<span><span class="co"># Run the efficient model</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps   pnew    pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.049<span style="text-decoration: underline;">9</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.050<span style="text-decoration: underline;">0</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2     1 0.049<span style="text-decoration: underline;">8</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     2 0.049<span style="text-decoration: underline;">8</span> 0.001<span style="text-decoration: underline;">00</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span>
<span></span>
<span><span class="co"># Quality control checks</span></span>
<span><span class="va">QC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_checks.html">QC_checks</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mutation rates in new reads looks good!</span></span>
<span><span class="co">#&gt; Background mutation rate looks good!</span></span>
<span><span class="co">#&gt; Average fraction news for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Exp_ID [2]</span></span></span>
<span><span class="co">#&gt;   Exp_ID Replicate avg_fn</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      1         1  0.312</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      1         2  0.315</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      2         1  0.325</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      2         2  0.321</span></span>
<span><span class="co">#&gt; The average fraction news in all samples are between 0.2 and 0.8, </span></span>
<span><span class="co">#&gt;               suggesting an appropriate label time!</span></span>
<span><span class="co">#&gt; logit(fn) correlations for each pair of replicates are:</span></span>
<span><span class="co">#&gt;   Exp_ID Rep_ID1 Rep_ID2 correlation</span></span>
<span><span class="co">#&gt; 1      1       1       2   0.8961197</span></span>
<span><span class="co">#&gt; 2      2       1       2   0.9262164</span></span>
<span><span class="co">#&gt; logit(fn) correlations are high, suggesting good reproducibility!</span></span>
<span><span class="co">#&gt; I suggest running the Hybrid implementation next. This can be done </span></span>
<span><span class="co">#&gt;               with bakRFit(Fit, HybridFit = TRUE), where Fit is your bakRFit object.</span></span></code></pre></div>
<p>QC_checks assesses four aspects of your data and bakR’s model
fit:</p>
<ol style="list-style-type: decimal">
<li>Raw mutation rates</li>
<li>Inferred new and old read mutation rates</li>
<li>Average fraction new</li>
<li>Fraction new estimate replicate correlation</li>
</ol>
<div class="section level3">
<h3 id="raw-mutation-rates">Raw mutation rates<a class="anchor" aria-label="anchor" href="#raw-mutation-rates"></a>
</h3>
<p>The first thing that <code><a href="../reference/QC_checks.html">QC_checks()</a></code> assesses is the average
mutation rate in all sequencing reads from all analyzed samples. High
quality data is characterized by mutation rates that are higher in all
+s<sup>4</sup>U samples than in any -s<sup>4</sup>U samples. In
addition, to a first order approximation the higher the mutation rates
in +s<sup>4</sup>U samples and the lower the mutation rates in
-s<sup>4</sup>U samples, the better. One output of QC_checks is a
convenient plot of raw mutation rates to see what these look like for
your data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Barplots of raw mutation rates</span></span>
<span><span class="va">QC</span><span class="op">$</span><span class="va">raw_mutrates</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In this simulated data, you can see that the +s<sup>4</sup>U raw
mutation rates are all consistenly much higher than the -s<sup>4</sup>U
mutation rates. The barplot also provides a quasi-arbitrary guide for
what good +s<sup>4</sup>U and -s<sup>4</sup>U mutation rates are.</p>
</div>
<div class="section level3">
<h3 id="inferred-mutation-rates">Inferred mutation rates<a class="anchor" aria-label="anchor" href="#inferred-mutation-rates"></a>
</h3>
<p>The second thing that <code><a href="../reference/QC_checks.html">QC_checks()</a></code> assesses is the
mutation rate of new and old reads, inferred by bakR. The difference
between the inferred and raw mutation rates can be a cause of confusion
for those new to NR-seq data. The raw mutation rates are the average
proportion of Ts that are mutated in all sequencing reads for a given
sample. This means that the raw mutation rates average over mutation
rates from both new reads (i.e., reads derived from RNA synthesized
during metabolic labeling) and old reads (i.e., those that were
synthesized prior to metabolic labeling). The inferred mutation rates
are the estimated mutation rates of old and new reads separately. Higher
inferred mutation rates yield higher raw mutation rates, but they are
distinct quantities. In particular, the raw mutation rate depends not
only on the new and old read mutation rates, but on what fraction of
sequencing reads are new and what fraction are old.</p>
<p>Like with the raw mutation rates, higher new read mutation rates and
lower old read mutation rates are better. <code><a href="../reference/QC_checks.html">QC_checks()</a></code>
outputs a helpful visualization of these mutation rates:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Barplots of inferred mutation rates</span></span>
<span><span class="va">QC</span><span class="op">$</span><span class="va">conversion_rates</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Once again, this simulated data looks good!</p>
</div>
<div class="section level3">
<h3 id="average-fraction-news">Average fraction news<a class="anchor" aria-label="anchor" href="#average-fraction-news"></a>
</h3>
<p>The third thing that <code><a href="../reference/QC_checks.html">QC_checks()</a></code> assesses is the average
estimated fraction new. The ability to accurately infer the degradation
kinetics of a particular RNA feature is highly dependent on the
relationship between the length of metabolic labeling and the half-life
of the feature. More specifically, some <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007252" class="external-link">really
nice work</a> from the Dieterich lab has shown that degradation rate
constant estimates are more accurate for features with half-lives closer
to the metabolic labeling time. If the label time is much longer
(shorter) than a feature’s half-live, than the feature will be almost
completely labeled (unlabeled). The relationship between fraction new
and degradation rate constant makes it so that at extreme fraction news
(close to 0 or 1), the rate constant estimate is very sensitive to the
exact fraction new estimate. This means that fraction news close to 0 or
1 yield more uncertain kinetic parameter estimates.</p>
<p>QC_checks() will output a message about the average estimated
fraction new in each +s<sup>4</sup>U sample. You can also visualize the
distribution of fraction new estimates as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Barplots of inferred mutation rates</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logit_fn</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_d</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"logit(fn) estimates"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"density"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Note, the density plot above is showing fraction news on a logit
scale. Logit is a function that takes as input numbers bounded between 0
and 1 and outputs an unbounded number. A logit(fraction new) of 0 means
a fraction new of 0.5. A logit(fraction new) of -2 (2) is a fraction new
of ~0.1 (~0.9) Similar to log-transforming RNA-seq read count data,
logit transforming fraction news can be useful when visualizing data
that ranges multiple orders of magnitude.</p>
</div>
<div class="section level3">
<h3 id="fraction-new-estimate-replicate-correlation">Fraction new estimate replicate correlation<a class="anchor" aria-label="anchor" href="#fraction-new-estimate-replicate-correlation"></a>
</h3>
<p>Finally, <code><a href="../reference/QC_checks.html">QC_checks()</a></code> assesses the extent to which
fraction new estimates correlate between replicates of the same
experimental condition. Better correlation means less variable data,
which means it will be easier to identify differences in kinetic
parameters between experimental conditions. <code>QC_checks</code> will
output a message regarding each replicate-to-replicate correlation, and
also provides a number of correlation plots that you can inspect:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Barplots of inferred mutation rates</span></span>
<span>  <span class="co"># Numerical indices are ordered as they appear in QC_checks() output message</span></span>
<span>  <span class="co"># So this is for replicate 1 and 2 of experimental ID 1</span></span>
<span><span class="va">QC</span><span class="op">$</span><span class="va">correlation_plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In addition, <code>QC_checks</code> outputs a correlation matrix to
allow for convenient visualization of all sample-to-sample
correlations:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using function from corrplot package</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.mixed.html" class="external-link">corrplot.mixed</a></span><span class="op">(</span><span class="va">QC</span><span class="op">$</span><span class="va">correlation_matrix</span>, </span>
<span>               upper <span class="op">=</span> <span class="st">"square"</span>, lower <span class="op">=</span> <span class="st">"number"</span>, </span>
<span>               addgrid.col <span class="op">=</span> <span class="st">"black"</span>, tl.col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="challenge-to-be-aware-of-inaccurate-pnew-and-pold-estimates">Challenge to be aware of: inaccurate pnew and pold estimates<a class="anchor" aria-label="anchor" href="#challenge-to-be-aware-of-inaccurate-pnew-and-pold-estimates"></a>
</h2>
<p><code>QC_checks</code> is designed to identify and flag challenges
that bakR could run into when analyzing your data. In my experience, the
single most important thing to check is the robustness of the new and
old read mutation rate estimates (pnew and pold).</p>
<p>By default, bakR fits a simple binomial mixture with the method of
maximum likelihood to estimate pnew. If you have -s<sup>4</sup>U data,
then bakR uses the average mutation rate from that data as a global pold
estimate. Otherwise, the default pold estimation strategy is the same as
for pnew. This strategy can go astray if most of the reads in your data
are either new or old. Intuitively, if there are very few new (old)
reads, it is very difficult for a model to infer what the mutation rate
in new (old) reads is. For example, here is an analysis of simulated
data where over 98% of reads are old:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">321</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate a nucleotide recoding dataset</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Simulate_bakRData.html">Simulate_bakRData</a></span><span class="op">(</span><span class="fl">1000</span>, nreps <span class="op">=</span> <span class="fl">2</span>, fn_mean <span class="op">=</span> <span class="op">-</span><span class="fl">4</span><span class="op">)</span></span>
<span>  <span class="co"># This will simulate 1000 features, 2 experimental conditions,</span></span>
<span>  <span class="co"># and 2 replicates for each experimental condition</span></span>
<span>  <span class="co"># The average logit(fn) will be -4, which corresponds to an average fn of just under 0.02.</span></span>
<span></span>
<span><span class="co"># Run the efficient model</span></span>
<span>  <span class="co"># Check the pnew estimates, which should all be around 0.05</span></span>
<span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="va">bakRData</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding reliable Features</span></span>
<span><span class="co">#&gt; Filtering out unwanted or unreliable features</span></span>
<span><span class="co">#&gt; Processing data...</span></span>
<span><span class="co">#&gt; Estimating pnew with likelihood maximization</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   mut [2]</span></span></span>
<span><span class="co">#&gt;     mut  reps    pnew    pold</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1     1 0.050<span style="text-decoration: underline;">5</span>  0.001<span style="text-decoration: underline;">01</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     1     2 0.049<span style="text-decoration: underline;">3</span>  0.001<span style="text-decoration: underline;">01</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     2     1 0.002<span style="text-decoration: underline;">21</span> 0.001<span style="text-decoration: underline;">01</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     2     2 0.002<span style="text-decoration: underline;">23</span> 0.001<span style="text-decoration: underline;">01</span></span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>While the pnew estimates in one of the experimental conditions is
alright, the estimates for the other condition are both massive
underestimates. Running <code>QC_checks</code> provides a suggestion
though</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run QC_checks and read messages</span></span>
<span><span class="va">QC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_checks.html">QC_checks</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in QC_checks(Fit): Mutation rates in new reads are below 0.7% in one or more samples. </span></span>
<span><span class="co">#&gt;               It is very difficult to identify kinetic differences with such low</span></span>
<span><span class="co">#&gt;               mutation rates.</span></span>
<span><span class="co">#&gt; Background mutation rate looks good!</span></span>
<span><span class="co">#&gt; Average fraction news for each sample are:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Exp_ID [2]</span></span></span>
<span><span class="co">#&gt;   Exp_ID Replicate avg_fn</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      1         1 0.046<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      1         2 0.047<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      2         1 0.533 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      2         2 0.531</span></span>
<span><span class="co">#&gt; Warning in assess_fn_cor(Fit, Bad_data = Bad_data, bakRFn = FALSE): The average fraction news are extremely low (less than 0.05) in </span></span>
<span><span class="co">#&gt;               one or more samples, suggesting your label time was too short. </span></span>
<span><span class="co">#&gt;               It will be difficult for bakR to identify any kinetic differences.</span></span>
<span><span class="co">#&gt; Low fraction news impair bakR's default mutation rate estimation </span></span>
<span><span class="co">#&gt;                 strategy. I suggest rerunning bakRFit with FastRerun and </span></span>
<span><span class="co">#&gt;                 StanRateEst = TRUE, particularly if some of the estimated </span></span>
<span><span class="co">#&gt;                 mutation rates are oddly low (&lt; 0.01) in a subset of samples.</span></span>
<span><span class="co">#&gt; logit(fn) correlations for each pair of replicates are:</span></span>
<span><span class="co">#&gt;   Exp_ID Rep_ID1 Rep_ID2 correlation</span></span>
<span><span class="co">#&gt; 1      1       1       2   0.8533296</span></span>
<span><span class="co">#&gt; 2      2       1       2   0.7852611</span></span>
<span><span class="co">#&gt; logit(fn) correlations are high, suggesting good reproducibility!</span></span>
<span><span class="co">#&gt; Some aspects of your data may limit bakR's ability to detect </span></span>
<span><span class="co">#&gt;               differential kinetics. Check warning messages for details.</span></span></code></pre></div>
<p>Note the message that says “I suggest rerunning bakRFit with
FastRerun and StanRateEst = TRUE, particularly if some of the estimated
mutation rates are oddly low (&lt; 0.01) in a subset of samples”. bakR
has a second pnew and pold estimation strategy up it’s sleeve,
accessible via the <code>StanRateEst</code> parameter. Setting this to
TRUE causes bakR to resort to using a fully Bayesian approach with the
probabilistic programming language <code>Stan</code> working on the back
end to fit a mixture model. This approach will take a bit longer to run,
but can provide more accurate pnew and pold estimates:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rerun with Stan-based pnew estimation</span></span>
<span>  <span class="co"># This will take a couple minutes to run</span></span>
<span><span class="va">Fit_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">Fit</span>, FastRerun <span class="op">=</span> <span class="cn">TRUE</span>, StanRateEst <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; SAMPLING FOR MODEL 'Mutrate_est' NOW (CHAIN 1).</span></span>
<span><span class="co">#&gt; Chain 1: </span></span>
<span><span class="co">#&gt; Chain 1: Gradient evaluation took 0.000475 seconds</span></span>
<span><span class="co">#&gt; Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 4.75 seconds.</span></span>
<span><span class="co">#&gt; Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">#&gt; Chain 1: </span></span>
<span><span class="co">#&gt; Chain 1: </span></span>
<span><span class="co">#&gt; Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1: </span></span>
<span><span class="co">#&gt; Chain 1:  Elapsed Time: 10.591 seconds (Warm-up)</span></span>
<span><span class="co">#&gt; Chain 1:                5.656 seconds (Sampling)</span></span>
<span><span class="co">#&gt; Chain 1:                16.247 seconds (Total)</span></span>
<span><span class="co">#&gt; Chain 1:</span></span>
<span><span class="co">#&gt; Estimating pnew with Stan output</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with Stan output</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt;   mut reps       pnew         pold</span></span>
<span><span class="co">#&gt; 1   1    1 0.04221830 0.0009098713</span></span>
<span><span class="co">#&gt; 2   1    2 0.04332409 0.0009098713</span></span>
<span><span class="co">#&gt; 3   2    1 0.04003344 0.0009098713</span></span>
<span><span class="co">#&gt; 4   2    2 0.03945663 0.0009098713</span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>It’s incredibly difficult to get perfect pnew estimates on this
dataset, but this strategy certainly does a lot better! You can compare
the resulting fraction new estimates in the troublesome samples:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Simulated ground truth</span></span>
<span><span class="va">sim_truth</span> <span class="op">&lt;-</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">sim_list</span></span>
<span></span>
<span><span class="co"># Features that made it past filtering</span></span>
<span><span class="va">XFs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Effects_df</span><span class="op">$</span><span class="va">XF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulated logit(fraction news) from features making it past filtering</span></span>
<span><span class="va">true_fn</span> <span class="op">&lt;-</span> <span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Logit_fn</span><span class="op">[</span><span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Feature_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">XFs</span> <span class="op">&amp;</span></span>
<span>                                           <span class="va">sim_truth</span><span class="op">$</span><span class="va">Fn_rep_sim</span><span class="op">$</span><span class="va">Exp_ID</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Estimated logit(fraction news)</span></span>
<span><span class="va">est_fn</span> <span class="op">&lt;-</span> <span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">logit_fn</span><span class="op">[</span><span class="va">Fit</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">Exp_ID</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Compare estimate to truth</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">true_fn</span>, <span class="va">est_fn</span>, xlab <span class="op">=</span> <span class="st">"True logit(fn)"</span>, ylab <span class="op">=</span> <span class="st">"Estimated logit(fn)"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Default pnew estimation"</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Estimated logit(fraction news)</span></span>
<span><span class="va">est_fn</span> <span class="op">&lt;-</span> <span class="va">Fit_s</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">logit_fn</span><span class="op">[</span><span class="va">Fit_s</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">Exp_ID</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Compare estimate to truth</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">true_fn</span>, <span class="va">est_fn</span>, xlab <span class="op">=</span> <span class="st">"True logit(fn)"</span>, ylab <span class="op">=</span> <span class="st">"Estimated logit(fn)"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Alternative pnew estimation"</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-11-2.png" width="576" style="display: block; margin: auto;"></p>
<p>Perhaps you have an alternative strategy for estimating pnew, or you
are fairly confident as to what the mutation rate in new reads in your
data is. In that case, your estimates can be passed directly to bakR no
questions asked. You can either pass a vector of pnews for each sample
(ordered as they appear in the bakRFit pnew/pold estimate message, so
ordered by Experimental ID then Replicate ID), or just a single pnew if
you think it is the same in all samples:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rerun with Stan-based pnew estimation</span></span>
<span>  <span class="co"># This will take a couple minutes to run</span></span>
<span><span class="va">Fit_u</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bakRFit.html">bakRFit</a></span><span class="op">(</span><span class="va">Fit</span>, FastRerun <span class="op">=</span> <span class="cn">TRUE</span>, pnew <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using provided pnew estimates</span></span>
<span><span class="co">#&gt; Estimating unlabeled mutation rate with -s4U data</span></span>
<span><span class="co">#&gt; Estimated pnews and polds for each sample are:</span></span>
<span><span class="co">#&gt;   mut reps pnew        pold</span></span>
<span><span class="co">#&gt; 1   1    1 0.05 0.001006472</span></span>
<span><span class="co">#&gt; 2   1    2 0.05 0.001006472</span></span>
<span><span class="co">#&gt; 3   2    1 0.05 0.001006472</span></span>
<span><span class="co">#&gt; 4   2    2 0.05 0.001006472</span></span>
<span><span class="co">#&gt; Estimating fraction labeled</span></span>
<span><span class="co">#&gt; Estimating per replicate uncertainties</span></span>
<span><span class="co">#&gt; Estimating read count-variance relationship</span></span>
<span><span class="co">#&gt; Averaging replicate data and regularizing estimates</span></span>
<span><span class="co">#&gt; Assessing statistical significance</span></span>
<span><span class="co">#&gt; All done! Run QC_checks() on your bakRFit object to assess the </span></span>
<span><span class="co">#&gt;             quality of your data and get recommendations for next steps.</span></span></code></pre></div>
<p>As expected, this improves estimate accuracy, though not much more
than setting <code>StanRateEst = TRUE</code> did:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Estimated logit(fraction news)</span></span>
<span><span class="va">est_fn</span> <span class="op">&lt;-</span> <span class="va">Fit_u</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">logit_fn</span><span class="op">[</span><span class="va">Fit_u</span><span class="op">$</span><span class="va">Fast_Fit</span><span class="op">$</span><span class="va">Fn_Estimates</span><span class="op">$</span><span class="va">Exp_ID</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Compare estimate to truth</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">true_fn</span>, <span class="va">est_fn</span>, xlab <span class="op">=</span> <span class="st">"True logit(fn)"</span>, ylab <span class="op">=</span> <span class="st">"Estimated logit(fn)"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"User provided pnew"</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Troubleshooting_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="experimental-suggestions">Experimental suggestions<a class="anchor" aria-label="anchor" href="#experimental-suggestions"></a>
</h2>
<p>Sometimes the solution to challenges you might face when analyzing
NR-seq data with bakR can be overcome with alternative bakR settings.
Sometimes though, experimental optimization can make a much bigger
difference. Here are some general suggestions to consider when designing
your next NR-seq experiment:</p>
<ol style="list-style-type: decimal">
<li>Assess and optimize s<sup>4</sup>U incorporation in your cell line.
<ul>
<li>You would be surprised how much cell line to cell line variability
there is in readiness of metabolic label uptake. On one end of the
spectrum there are HeLa cells that uptake label so ferociously that you
often can use lower than normal s<sup>4</sup>U concentrations to avoid
cytotoxic effects while still maintaining high incorporation rates. On
the other end of the spectrum though, neuronal cell lines often uptake
label at a much lower rate.</li>
<li>TAMRA dot blot assays are one way our lab inspects the incorporation
rate in a new cell line. If incorporation rates are low, we test using a
higher concentration of metabolic label, while looking out for any signs
of cytotoxicity</li>
</ul>
</li>
<li>
<strong>Don’t user super short label times!!</strong>
<ul>
<li>There is a tendency to treat NR-seq experiments like transcriptional
assays. The thinking goes “label for a short enough time and the amount
of new RNA is only influenced by synthesis kinetics”.</li>
<li>This is fraught for many reasons. For one, a “short” label time in
an NR-seq experiment still often means about an hour in mammalian cell
lines, as enough RNA has to be labeled to allow for robust detection of
it without enrichment of the labeled species. There is plenty of
degradation of new RNA in that time frame though, meaning that the
amount of new RNA is a function of both degradation and synthesis
kinetics.</li>
<li>In addition, by restricting yourself to analysis of only new RNA,
you are throwing out the vast majority of your reads and significantly
underpowering your analyses.</li>
<li>Note, while there is some evidence in the literature that
differential new RNA expression analysis can provide insights missed by
differential total RNA expression analysis, the extent to which this is
generally true is questionable. Evidence in favor of this conclusion
could stem from inadequate normalization of the total RNA-seq data,
causing global effects of a perturbation to be missed.</li>
<li>Finally, one benefit of analyzing the fraction of reads that are new
rather than the raw number of reads that are new is that the former is
“internally normalized” The fraction new is the ratio of an RNA
feature’s new read counts in a sample to it’s total read counts in that
same sample. Thus, any scale factor that you need to apply to the
sample-wide or feature-specific read counts cancel out when calculating
this ratio. Analyzing fraction news (e.g., by relating these to
degradation rate constants) thus avoids the challenge of normalizing
RNA-seq read count data. This upside is eliminated when performing
differential expression analysis with new RNA read counts, where
normalization must once again be considered.<br>
</li>
</ul>
</li>
<li>Rather, try and use a label time near the average half-life of RNAs
that you are interested in studying. If you are hoping to probe all
mRNAs, then note that in humans the average mRNA half-life is around 4
hours. If you are studying miRNAs, which have much longer half-lives,
you will likely want to use a longer label time.
<ul>
<li>Beware of toxicity effects of the metabolic label though!</li>
<li>Always include a -s<sup>4</sup>U control so that you can determine
if there is anything significantly different about expression patterns
in s<sup>4</sup>U treated cells (e.g., activation of stress response
pathways). Be careful not to overinterpret effects like metabolic label
containing RNA dropout (discussed in the “Correcting for dropout”
vignette) as cytotoxicity effects! Correct read counts for dropout (with
bakR’s <code>CorrectDropout</code> function) and see if there are still
major expression differences</li>
</ul>
</li>
<li>Finally, I would suggest avoiding pulse-chase experimental designs,
instead opting for a pulse-label design
<ul>
<li>For clarity, pulse-chase = pulse with s<sup>4</sup>U then chase with
uridine; pulse-label = pulse with s<sup>4</sup>U.</li>
<li>Pulse-chase experimental designs suffer from a number of
shortcomings and are almost never necessary in NR-seq experiments.</li>
<li>There are many classic examples of pulse-chase experimental designs
being used to assess the kinetics of RNA metabolism. In these cases
though, the point of the pulse was to create a species of RNA whose
dynamics are completely driven by the degradation kinetics of the RNA.
For example, the pioneering studies from the Parker lab elucidating the
mechanisms of RNA degradation involved pulsing with a particular
nutrient to stimulate transcription of a construct. The chase then
washed out that nutrient to shut off transcription from the construct,
meaning that RNA produced from it will exclusively degrade
post-chase.</li>
<li>When you pulse with a metabolic label like s<sup>4</sup>U though,
you have already created a species of RNA whose dynamics are completely
degradation driven: the unlabeled RNA. Following the metabolic label
pulse with a nucleotide chase is thus redundant, switching the
degradation driven population from the unlabeled RNA to the labeled
RNA.</li>
<li>In addition, pulse-chase experiments often necessitate extensive
exposure of your cells to the metabolic label, which increases the
possibility of cytotoxic effects of metabolic label.</li>
<li>Finally, analysis of pulse-chase experiments is more complicated
than a pulse-label design. To estimate the kinetics of degradation, you
need to know what fraction of the RNA for each feature of interest was
labeled after the pulse. This means that estimating the synthesis and
degradation kinetics requires two separate samples, RNA extracted after
the pulse and RNA extracted after the chase. With the pulse alone, you
can get all of the information that you can get with both though!! You
want multiple time points? Just do multiple pulses. Pulse-labels are
easier to perform and easier to analyze than pulse-chases.</li>
<li>If you have pulse-chase NR-seq data you would like to analyze, you
can contact me by either posting an Issue to the bakR Github or emailing
me directly (<a href="mailto:isaac.vock@yale.edu" class="email">isaac.vock@yale.edu</a>). I also have a <a href="https://github.com/isaacvock/Pulse_Chase" class="external-link">repo</a> that was
created for this purpose you can check out.</li>
</ul>
</li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Isaac Vock.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
